---
title: "R Notebook"
output: html_notebook
---

#Allele imbalance analysis + brief comparison to WES variants


Import libraries
```{r}
suppressPackageStartupMessages(library(VariantAnnotation))
suppressPackageStartupMessages(library(tidyr))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(tibble))
suppressPackageStartupMessages(library(TxDb.Hsapiens.UCSC.hg19.knownGene))
suppressPackageStartupMessages(library(SNPlocs.Hsapiens.dbSNP.20101109))
suppressPackageStartupMessages(library(GenomicRanges))
suppressPackageStartupMessages(library(org.Hs.eg.db))
suppressPackageStartupMessages(library(PMCMR))
```

Functions
```{r}
#read in VCF, reformat and calculate allelic expression values
AllelicExpressionFromVCF<- function(File, GOIRanges, TotalCoverage, RefCoverage, AltCoverage, NVariantsPerGene, QualityThreshold){
  #read in data
  vcf<- read.table(File)
  vcf$V9<- lapply(vcf$V9, GetMaxofWeirdList) #Calls function- for multi alt alleles, gets max alt depth
  vcf$V9<- as.integer(vcf$V9)
  colnames(vcf)<- c("seqnames", "start", "ID", "Ref", "Alt", "Score", "Filter", "RefDepth", "AltDepth")
  vcf$end<- vcf$start

  #convert to granges object
  vcf_gr<- makeGRangesFromDataFrame(vcf, keep.extra.columns = TRUE)

  #Get overlaps of GOI ranges and all SNPs
  ranges<- subsetByOverlaps(vcf_gr, GOIRanges)
    
  #Get the geneID names, associate back with the overlap GRanges object
  hits<- findOverlaps(vcf_gr, GOIRanges)
  geneid <- CharacterList(split(GOIRanges$gene_id[subjectHits(hits)],queryHits(hits)))
  mcols(ranges) <- DataFrame(mcols(ranges), geneid)
  Junctions_GeneIDs<- as.data.frame(ranges)
  
  #Calculate minor allele ratio and allelic expression (abs(0.5-minor allele ratio))
  Junctions_GeneIDs$Coverage<- rowSums(Junctions_GeneIDs[,11:12])
  Junctions_GeneIDs$MAF<- Junctions_GeneIDs$AltDepth/Junctions_GeneIDs$Coverage
  Junctions_GeneIDs$AllelicExpression<- abs(0.5-Junctions_GeneIDs$MAF)
  
  #Subset by SNPs that meet coverage/quality thresholds
  TotalCoverageFilter<- Junctions_GeneIDs[Junctions_GeneIDs$Coverage>=TotalCoverage,]
  RefCoverageFilter<- TotalCoverageFilter[TotalCoverageFilter$RefDepth>=RefCoverage,]
  AltCoverageFilter<- RefCoverageFilter[RefCoverageFilter$AltDepth>=AltCoverage,]
  AltCoverageFilter<- AltCoverageFilter[AltCoverageFilter$Score >=QualityThreshold,]
  
  #subset by genes with at least n variants
  AltCoverageFilter$geneid<-as.character(AltCoverageFilter$geneid)
  NVariants<- AltCoverageFilter%>% group_by(geneid) %>% mutate(NumberVariantsPerGene=n()) %>% as.data.frame()
  NVariantsFilter<- NVariants[NVariants$NumberVariantsPerGene>=NVariantsPerGene,]
    
  #median of these ratios for genes
  MedianRatios<- NVariantsFilter %>% group_by(geneid) %>% mutate(MedianAllelicExpression=median(AllelicExpression)) %>% as.data.frame
  MedianRatios<- MedianRatios[order(MedianRatios$geneid),]
  
  
  return(MedianRatios)
  
}

#Add ID unique to that SNP (chr, location, ref/alt)
AddUniqueSNPID<- function(df){
  df$SNPID<- paste(df$seqnames, ":", df$start, "_", df$Ref, "/", df$Alt, sep="")
  return(df)
}

#Recalculate allelic expression after performing other filtering
NVariantsFilter_AllelicExpresion<- function(df, NVariantsPerGene){
  #subset by genes with at least n variants
  NVariants<- df%>% group_by(geneid) %>% mutate(NumberVariantsPerGene_AfterExome=n()) %>% as.data.frame()
  NVariantsFilter<- NVariants[NVariants$NumberVariantsPerGene_AfterExome>=NVariantsPerGene,]
    
  #median of these ratios for genes
  MedianRatios<- NVariantsFilter %>% group_by(geneid) %>% mutate(MedianAllelicExpression=median(AllelicExpression)) %>% as.data.frame
  MedianRatios<- MedianRatios[order(MedianRatios$geneid),]
  
  return(MedianRatios)
}

GetMaxofWeirdList<- function(WeirdList){
  return(max(as.integer(unlist(strsplit(as.character(WeirdList), ",")))))
}
```


#Read in GOI, associate with Geneids and create Granges object
(For reading in exome data, getting geneids later)
```{r}
#read in txdb+genes
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
genes <- as.data.frame(genes(TxDb.Hsapiens.UCSC.hg19.knownGene))
#subset list of all genes (with positions) by genes of interest. convert to Granges: CM genes
geneNames<- read.csv("//Cifs2/rcdata$/UKRNASeq/VariantCalling/GenesofInterest.csv",header=FALSE)
GOI<- unname(mapIds(org.Hs.eg.db, keys=as.character(geneNames$V1), column="ENTREZID",keytype="SYMBOL"))
GOIRanges<- makeGRangesFromDataFrame(genes[rownames(genes) %in% GOI,], keep.extra.columns=TRUE)
#subset list of all genes (with positions) by genes of interest. convert to Granges: non CM genes
geneNames<- read.csv("//Cifs2/rcdata$/UKRNASeq/VariantCalling/NonCM_GenesofInterest.csv",header=FALSE)
GOI<- unname(mapIds(org.Hs.eg.db, keys=as.character(geneNames$V1), column="ENTREZID",keytype="SYMBOL"))
NonCM_GOIRanges<- makeGRangesFromDataFrame(genes[rownames(genes) %in% GOI,], keep.extra.columns=TRUE)
```

#Read in data- RNASeq and Exome variants for SNPs in CM and NonCM gene list


Some preliminary filtering here:
  1. >=15X total coverage
  2. >=2x Ref and Alt coverage 
  3. >=3 variants in a gene (for RNASeq)
  4. >=40 qual score (for Exome)
  
Minor allele frequence and "Allelic Expression" (=abs(0.5-MAF)) also calculated here for all variants that pass these filters

#Read in RNASeq vcf data: 3 sources:

1) Merged alignment (Samples 1-9,11-13,15-19)
2) Run 1 (Samples 10 and 14)
3) Run 2 (Controls)


```{r}
extension<- ".vcf.gz.txt"

############
#Dataset 1 #
############

#CM variants
        #set list of files- merged files
        merged_readwd<- "D:/UKy/MergingRNASeq/MergedRuns_vcf.gz/CM_GOIFiltered/"
        merged_samples<- c(1:9,11:13,15:19)
        merged_files<-c(paste(merged_readwd,"AllelicDepthFinal_GOIFilteredvcf", merged_samples, "_Merged", extension,sep=""))
        
        #Read in vcf data, get Allelic expression stats, preliminary filtering
        merged_AllCMvcf<- lapply(merged_files, AllelicExpressionFromVCF, GOIRanges=GOIRanges, 15,2,2,0,0)
        
#NonCM variants
        #set list of files- merged files
        merged_readwd<- "D:/UKy/MergingRNASeq/MergedRuns_vcf.gz/NonCM_GOIFiltered/"
        merged_samples<- c(1:9,11:13,15:19)
        merged_files<-c(paste(merged_readwd,"NonCM_AllelicDepthFinal_NonCM_GOIFilteredvcf", merged_samples, "_Merged", extension,sep=""))
        
        #Read in vcf data, get Allelic expression stats, preliminary filtering
        merged_AllNonCMvcf<- lapply(merged_files, AllelicExpressionFromVCF, GOIRanges=NonCM_GOIRanges, 15,2,2,0,0)
        
#Merge CM and NonCM variants into one list
merged_Allvcf<-Map(rbind, merged_AllCMvcf, merged_AllNonCMvcf)

############
#Dataset 2 #
############

#CM variants
        #set list of files
        run1_readwd<- "D:/UKy/RNASeq_RawVcf/Run1_vcfs/"
        run1_samples<- c(10,14)
        run1_files<-c(paste(run1_readwd,"AllelicDepthFinal_GOIFilteredvcf", run1_samples, extension,sep=""))
        
        #Read in vcf data, get Allelic expression stats, preliminary filtering
        run1_AllCMvcf<- lapply(run1_files, AllelicExpressionFromVCF, GOIRanges=GOIRanges, 15,2,2,0,0)
        
#NonCM variants
        #set list of files
        run1_readwd<- "D:/UKy/RNASeq_RawVcf/Run1_vcfs/"
        run1_samples<- c(10,14)
        run1_files<-c(paste(run1_readwd,"NonCM_AllelicDepthFinal_NonCM_GOIFilteredvcf", run1_samples, extension,sep=""))
        
        #Read in vcf data, get Allelic expression stats, preliminary filtering
        run1_AllNonCMvcf<- lapply(run1_files, AllelicExpressionFromVCF, GOIRanges=NonCM_GOIRanges, 15,2,2,0,0)

#Merge CM and NonCM variants into one list
Run1_Allvcf<-Map(rbind, run1_AllCMvcf, run1_AllNonCMvcf)

############
#Dataset 3 #
############


#CM variants
        #set list of files
        run2_readwd<- "D:/UKy/RNASeq_RawVcf/CMGenesROI_AllelicDepthFromVCF/"
        run2_samples<- paste("UK-Control-", 1:5, sep="")
        run2_files<-c(paste(run2_readwd, "AllelicDepthFinal_GOIFilteredvcf", run2_samples, extension,sep=""))
        
        #Read in vcf data, get Allelic expression stats, preliminary filtering
        run2_AllCMvcf<- lapply(run2_files, AllelicExpressionFromVCF, GOIRanges=GOIRanges, 15,2,2,0,0)
        
#NonCM variants
        #set list of files
        run2_readwd<- "D:/UKy/RNASeq_RawVcf/NonCMGenesROI_AllelicDepthFromVCF/"
        run2_samples<- paste("UK-Control-", 1:5, sep="")
        run2_files<-c(paste(run2_readwd,"AllelicDepthFinal_GOIFilteredvcf", run2_samples, extension,sep=""))
        
        #Read in vcf data, get Allelic expression stats, preliminary filtering
        run2_AllNonCMvcf<- lapply(run2_files, AllelicExpressionFromVCF, GOIRanges=NonCM_GOIRanges, 15,2,2,0,0)
        
#Merge CM and NonCM variants into one list
Run2_Allvcf<-Map(rbind, run2_AllCMvcf, run2_AllNonCMvcf)


#merge all runs into 1 list- all 24 samples
AllAllelicExpression<- c(merged_Allvcf, Run1_Allvcf, Run2_Allvcf)
```





#Read in exome data
```{r}
#Some preliminary filtering included here: 2x Ref and Alt coverage + >= 40 qual score

#Read in for exome vcf data- CM genes
wd<-"D:/UKy/Exome_Vcf/CMGenesROI_AllelicDepth/" #Set working directory
Samples<- c(1:3,5:19) #Samples to be read in
Files<- paste(wd, "AllelicDepth_GOIFilteredvcf_HeaderRemoved_Tube-", Samples, "-clean.snp.txt",sep="") #set file names
CM_AllAllelicExpression_Exome<-lapply(Files, AllelicExpressionFromVCF, GOIRanges=GOIRanges, 0,0,0,0,0)

#Read in for exome vcf data- Non CM genes
wd<-"D:/UKy/Exome_Vcf/NonCMGenesROI_AllelicDepth/" #Set working directory
Samples<- c(1:3,5:19) #Samples to be read in
Files<- paste(wd, "AllelicDepth_GOIFilteredvcf_HeaderRemoved_Tube-", Samples, "-clean.snp.txt",sep="") #set file names
NonCM_AllAllelicExpression_Exome<-lapply(Files, AllelicExpressionFromVCF, GOIRanges=NonCM_GOIRanges, 0,0,0,0,0)

#Merge CM and nonCM into one list
AllAllelicExpression_Exome<-Map(rbind, CM_AllAllelicExpression_Exome, NonCM_AllAllelicExpression_Exome) #26259 variants total
```

#Export (mainly) unfiltered exome variants
```{r}
#Add sample name
AllExome<-list()
for (i in 1:length(AllAllelicExpression_Exome)){
  df<-AllAllelicExpression_Exome[[1]]
  Sample<-Samples[i]
  df$Sample<-Sample
  AllExome[[i]]<-df
}

merge_Exome<-do.call(rbind, AllExome)

write.csv(merge_Exome, "C:/Users/jenna/OneDrive/Documents/UKy/AllExomeVariants_NoFilter.csv")
```

#For CM samples: what percent of exome variants were also identified in RNASeq variant calling? And vice versa?
```{r}
#Remove samples 10 and 14 from exome
Exome_Comparison<-AllAllelicExpression_Exome[!(Samples %in% c("10", "14"))]

#Remove controls + sample 4 from RNASeq
RNASamples<-c(merged_samples, run1_samples, run2_samples)
RNA_Comparison<-AllAllelicExpression[!(RNASamples %in% c("UK-Control-1", 
                                                    "UK-Control-2", 
                                                    "UK-Control-3", 
                                                    "UK-Control-4", 
                                                    "UK-Control-5",
                                                    "4", "10", "14"))]

#cycle through all 16 samples, compare which SNPs match
for (sample in 1:length(RNA_Comparison)){
  RNA<-RNA_Comparison[[sample]]
  DNA<-Exome_Comparison[[sample]]
  RNA$Name<-paste(RNA$seqnames, ":",RNA$start ,"-", RNA$end, "_", RNA$Ref, "/", RNA$Alt,sep="")
  DNA$Name<-paste(DNA$seqnames, ":",DNA$start ,"-", DNA$end, "_", DNA$Ref, "/", DNA$Alt,sep="")
  RNAinDNA<-sum(RNA$Name %in% DNA$Name)
  print(paste("Sample", sample, ":", RNAinDNA, "of", length(RNA$Name), "DNA SNPs found in RNA"))
}
```

#Allelic imbalance analysis


#Overall: Filter variants --> calculate allelic expression --> test for significantly different allelic imbalance compared to controls --> write all RNASeq and Exome variants that occur in significant gene/sample pairs


#Step 1:
#Filtering RNASeq variants 

Strategy: 
  1. Filter based on Exome variants
    A: combine all variants (named as seqnames:start_Ref/Alt) into one list 
    B: collapse into list with just unique variants
    C: filter all RNASeq variants (by sample) by this exome list
  2. Filter based on coverage
    A. Remove any variants if the Ref coverage is less than 10% of the total coverage of the variant
    B. Remove any variants if the Alt coverage is less than 10% of the total coverage of the variant

Then: 
  1. recalculate allelic expression
  2. Calculate median allelic expression values for each gene/sample
  
Filterd variants and median gene allelic expression values exporte

  

```{r}
#Filter Step 1: exome variants
  #Add column to all variants (in exome and RNASeq data) with unique SNP ID
  Exome_UniqueSNPID<-lapply(AllAllelicExpression_Exome, AddUniqueSNPID)
  RNA_UniqueSNPID<- lapply(AllAllelicExpression, AddUniqueSNPID)
  
  #Combine all variants into one list
  Exome_ScoreSNP<- lapply(Exome_UniqueSNPID, function(x) {x[,c(9,19)]})
  multi_full<- do.call(rbind,Exome_ScoreSNP)
  
  #Get just unique variants (i.e., get rid of duplicates). Keep the SNP with the highest score (for if I ever want to do extra filtering based on score)
  Exome_AllSNPs <- multi_full[order(multi_full[,'SNPID'],-multi_full[,'Score']),]
  Exome_AllSNPs<- Exome_AllSNPs[!duplicated(Exome_AllSNPs$SNPID),] #6999 unique SNPs in controls
  
  #Filter RNASeq variants based on filtered exome variants
  RNA_VariantsInExome<- lapply(RNA_UniqueSNPID, function(x) {x[x$SNPID %in% Exome_AllSNPs$SNPID,]}) #4810 --> 310 variants in each sample (on average)

#Filter step 2: Alt/Ref coverage by PERCENTAGE OF TOTAL COVERAGE. This is in addition to the raw thresholds set when importing data
  PercentThreshold<- 0.1
  RNA_VariantsInExome_Ref<- lapply(RNA_VariantsInExome, function(x){
    x[x$RefDepth>=round(x$Coverage*PercentThreshold),]
  })
  RNA_VariantsInExome_Alt<- lapply(RNA_VariantsInExome_Ref, function(x){
    x[x$AltDepth>=round(x$Coverage*PercentThreshold),]
  }) #down to 300 variants in each sample (on average)

#Recalculate median allelic expression values based on this new dataset of variants- need to re-filter by nVariants/gene
RNA_Subset_AfterExome<- lapply(RNA_VariantsInExome_Alt, function(x) {x[,c(1:16,19)]})
RNA_ReFilterAfterExome<- lapply(RNA_Subset_AfterExome, NVariantsFilter_AllelicExpresion, NVariantsPerGene=3) #down to 230 variants per sample (on average)


#Add columns for mean allelic expression + SD Allelic Expression
RNA_AddMean<- lapply(RNA_ReFilterAfterExome, function(x){
  x %>% group_by(geneid) %>% mutate(MeanAllelicExpression=mean(AllelicExpression)) %>% as.data.frame
})

RNA_AddMeanAndSD<- lapply(RNA_AddMean, function(x){
  x %>% group_by(geneid) %>% mutate(SDAllelicExpression=sd(AllelicExpression)) %>% as.data.frame
})

#Merge filtered variants for writing to csv
Samples<- c(merged_samples, run1_samples, run2_samples)
for (i in 1:length(RNA_AddMeanAndSD)){
  RNA_AddMeanAndSD[[i]]$Sample<- Samples[i]
}
FilteredVariants<-do.call(rbind, RNA_AddMeanAndSD)

#Combine all median allelic expression values for each geneid
RNA_MedianFinal<- lapply(RNA_AddMeanAndSD, function(x) {x[!duplicated(x$geneid),]})
RNA_MedianFinal_rbind<-do.call(rbind, RNA_MedianFinal)


#bind all median allelic expression values into 1 (keepin just SNP info + Sample + MedianAllelicExpression + gene), then write to csv
MediansAll<-  do.call(rbind, RNA_MedianFinal)
MediansAll<- MediansAll[order(MediansAll$geneid),colnames(MediansAll) %in% c("geneid","NumberVariantsPerGene", "MedianAllelicExpression", "MeanAllelicExpression", "SDAllelicExpression", "Sample")]
```

#Export results
```{r}
#Filtered variants- RNASeq and Exome
write.csv(FilteredVariants, paste(merged_readwd, "MERGED_FinalFilteredVariants_PercentCoverageExome_AllCMVariants.csv",sep=""))

write.csv(RNA_MedianFinal_rbind, paste(merged_readwd,"MERGED_AllGenes_AllFilteredRNASeqVariants_CoveragePer.1_Total15_RefAlt2_3PerGene_FilteredByExomeQual40Total15RefAlt2.csv", sep="" ))

#Median allelic expression for each gene
write.csv(MediansAll, paste(merged_readwd, "MERGED_CMGenes_MedianAllelicExpression_AllRNASeqVariants_CoveragePer.1_Total15_RefAlt1_3PerGene_FilteredByExomeQual40Total15RefAlt2.csv", sep=""))
```


#Alternative: using results of ASEReadCounter (just sample 2)
No exome filtering
```{r}
#Read in data, export in correct formating
ase_2<-read.table("C:/Users/jenna/OneDrive/Documents/UKy/ASEReadCounterOutput_Sample2.txt")
ase_2<-ase_2[-1,]
ase_2<-ase_2[,c(1,2,3,4,5,9,10,6,7)]

write.table(ase_2,"C:/Users/jenna/OneDrive/Documents/UKy/ASEReadCounterOutput_Reformated_Sample2.txt")
test<-read.table(file)

#Read back in, filter by GOI, get correct 
file<-"C:/Users/jenna/OneDrive/Documents/UKy/ASEReadCounterOutput_Reformated_Sample2.txt"

test_CM<-AllelicExpressionFromVCF(file, GOIRanges, 15, 2, 2,3,0)
test_NonCM<-AllelicExpressionFromVCF(file, NonCM_GOIRanges, 15, 2, 2,3,0)
test<-rbind(test_CM, test_NonCM) #9,463 variants

#Add columns required for sig test
test$Sample<-"2"
test$SNPID<- paste(test$seqnames, ":", test$start, "_", test$Ref, "/", test$Alt,sep="")


#Merge with filtered controls (filtered according to standard method, not with ASEReadCounter)
Controls<-FilteredVariants[grep("Control",FilteredVariants$Sample),c(1:16,18,19,22,17)]
colnames(Controls)<-gsub("_AfterExome", "", colnames(Controls))
test<-rbind(test, Controls)

FilteredVariants<-test
```

#Test allelic expression differences for significance
Using Dunn sig test (ANNOVA post hoc)

```{r}
#Remove indels- incorrect alignment for these variants common, don't want to calculate significance using these SNPs
nChar<- apply(FilteredVariants, 2, nchar)[,7:8]
All_NoIndels<-FilteredVariants[rowSums(nChar)==2,]

#Subset data- just need geneid, sample, and allelic expression
All_sigTest<- All_NoIndels[,colnames(All_NoIndels) %in% c("geneid", "AllelicExpression", "Sample")]

#Rename Control-1/2/3/4/5 to control- will use all 5 controls as 1 "sample"
for (i in 1:5){
  sub<-paste("UK-Control-", i, sep="")
  All_sigTest$Sample<- gsub(sub, "Control", All_sigTest$Sample)
}

#Convert to factor
All_sigTest$Sample<-as.factor(All_sigTest$Sample)
All_sigTest$Sample<- ordered(All_sigTest$Sample, levels= c("Control",1:9,11:13,15:19)) #for all samples
All_sigTest$Sample<- ordered(All_sigTest$Sample, levels= c("Control",2)) #for just sample 2

#Split dataframe into a list of separate dfs for each gene
SplitByGene<- split(All_sigTest, f=All_sigTest$geneid) #138 genes, 386 for ASEReadCounter, just sample 2

#only keep gene if there's at least 1 variant in a non-control sample + 3 control samples
#At least 3 variants in a control:
NControls<- unlist(lapply(SplitByGene, function(x) {
  length(grep("Control", x$Sample))
}))
EnoughControls<-SplitByGene[NControls>=3] #down to just 84 for ASEReadCounter, sample 2

#At least 1 CM sample:
NSamples<- lapply(EnoughControls, function(x) { x %>% filter(Sample!= "Control") %>% group_by(Sample) %>% mutate(NumberVariants=n()) %>%  filter(NumberVariants>=1) %>% as.data.frame})
EnoughSamples<- EnoughControls[unlist(lapply(NSamples,nrow))>=1] #down to 61 genes


#initialize empty list for appending significant gene/samples to
SigAllelicExpression<- data.frame()
AllAlleicExpression_Dunn<-data.frame()

#Run dunn test for all genes 
for (i in 1:length(EnoughSamples)){
  variants<-EnoughSamples[[i]]
  dunn<-with(variants, posthoc.kruskal.dunn.test(x=AllelicExpression, g=Sample, p.adjust.method="none"))
  pvalue<-as.data.frame(dunn$p.value[,1])
  colnames(pvalue)<- "Pvalue"
  pvalue$Sample<-rownames(pvalue)
  pvalue$gene<- variants$geneid[1]
  sig<-pvalue[pvalue$Pvalue<0.05,]
  SigAllelicExpression<- rbind(SigAllelicExpression, sig)
  AllAlleicExpression_Dunn<-rbind(AllAlleicExpression_Dunn, pvalue)
  
}

#Add on info about median/mean/sd of the significantly different genes, for the sig sample + across controls
SigAllelicExpression$UniqueID<- paste(SigAllelicExpression$Sample, SigAllelicExpression$gene, sep=":")
MediansAll$UniqueID<- paste(MediansAll$Sample, MediansAll$geneid, sep=":")
SigGenes_Medians<- MediansAll[MediansAll$UniqueID %in% SigAllelicExpression$UniqueID,]


#Add in controls-mean/median/SD of allelic expression across all 5 control samples
#can't just take mean of these metrics across the 5 samples- there's different n of variants supporting these metrics
#Need to weight each variant equally
AllControls_AEMetrics<- All_sigTest[grep("Control", All_sigTest$Sample),]
AllControls_AEMetrics<- AllControls_AEMetrics %>% group_by(geneid) %>% mutate(Controls_MedianAllelicExpression=median(AllelicExpression),
                                                                              Controls_MeanAllelicExpression=mean(AllelicExpression),
                                                                              Controls_SDAllelicExpression= sd(AllelicExpression)) %>% as.data.frame
AllControls_AEMetrics<- AllControls_AEMetrics[!(duplicated(AllControls_AEMetrics$geneid)),c(1,4:6)]
AllControls_SigGenes<- AllControls_AEMetrics[AllControls_AEMetrics$geneid %in% SigGenes_Medians$geneid,]
                                                                      
#Merge sig gene Allelic expression values from CM samples with corresponding control values 
ControlsandCM<- full_join(SigGenes_Medians, AllControls_SigGenes, by="geneid")

#Merge this with significance info- pvalue from dunn test
Sig_AndAllelicExpressionMetrics_id<- cbind(SigAllelicExpression, ControlsandCM, by="UniqueID")
Sig_AndAllelicExpressionMetrics<- Sig_AndAllelicExpressionMetrics_id[,c(4,3,2,1,6:9,11:13)]


#Prep for export: all the SNPs that caused the significant gene/sample pairings. Will merge this with the above table into one file manually in excel.
AllFilteredSNPs<- All_NoIndels
AllFilteredSNPs$UniqueID<- paste(AllFilteredSNPs$Sample, AllFilteredSNPs$geneid,sep=":")
AllFilteredSNPs_SigGeneSamples<- AllFilteredSNPs[AllFilteredSNPs$UniqueID %in% Sig_AndAllelicExpressionMetrics_id$UniqueID,]

#Add in the number of variants in significant sample/gene pairings- for individual SNPs
AllFilteredSNPS_NumberVariantsPer<-AllFilteredSNPs %>% group_by(UniqueID) %>% mutate(NumberVariantsPerGene=n()) %>% as.data.frame()
AllFilteredSNPS_NumberVariantsPer<-AllFilteredSNPS_NumberVariantsPer[,23:24]
AllFilteredSNPs_SigGeneSamples<- merge(AllFilteredSNPS_NumberVariantsPer,AllFilteredSNPs_SigGeneSamples, by="UniqueID")

#Add in the number of variants in sig gene/pairings- for overall gene/sample table
Sig_AndAllelicExpressionMetrics<-merge(Sig_AndAllelicExpressionMetrics, AllFilteredSNPS_NumberVariantsPer[!(duplicated(AllFilteredSNPS_NumberVariantsPer)),],by="UniqueID")
```

Export results
```{r}
#significant gene/sample pairings- no info on indv SNPs
writewd<- "//Cifs2/rcdata$/UKRNASeq/VariantCalling/"
write.csv(Sig_AndAllelicExpressionMetrics, paste(writewd, "MERGED_DunnSigTest_AllelicExpression_CMGenes_PerCoverageFilter.1.csv", sep=""))

#All SNPs that cause sig gene/sample pairings
write.csv(AllFilteredSNPs_SigGeneSamples, paste(writewd, "MERGED_RNASeqVariantsInCMGenes_CausingSigAllelicExpression.csv", sep=""))
```

#For genes/sample pairs of interest (i.e., sig Allelic imbalance), get ALL EXOME VARIANTS

```{r}
#Add sample info
for (i in 1:length(AllAllelicExpression_Exome)){
  AllAllelicExpression_Exome[[i]]$Sample<- Samples[i]
}

#Merge all exome variants together
ExomeVariants<- do.call(rbind, AllAllelicExpression_Exome)
ExomeVariants$UniqueID<- paste(ExomeVariants$Sample, ExomeVariants$geneid, sep=":")

#Filter exome variants- only keep variants that are found in a gene/sample pair that had sig allelic imbalance
SigAI_ExomeVariants<- ExomeVariants[ExomeVariants$UniqueID %in% SigAllelicExpression$UniqueID,]
```

Export
```{r}
#All exome variants in sig gene/sample pairs
write.csv(SigAI_ExomeVariants, paste(writewd, "MERGED_ExomeVariants_CMGenes_InGeneSamplePairSigAllelicImbalance.csv", sep=""))
```



#No longer in use! NonCM genes analysis combined with CM (above)

##################
#For Non CM genes#
##################

```{r}
#Filter Step 1: exome variants
  #Add column to all variants (in exome and RNASeq data) with unique SNP ID
  Exome_UniqueSNPID<-lapply(NonCM_AllAllelicExpression_Exome, AddUniqueSNPID)
  RNA_UniqueSNPID<- lapply(NonCM_AllAllelicExpression, AddUniqueSNPID)
  
  #Combine all variants into one list
  Exome_ScoreSNP<- lapply(Exome_UniqueSNPID, function(x) {x[,c(9,20)]})
  multi_full<- do.call(rbind,Exome_ScoreSNP)
  
  #Get just unique variants (i.e., get rid of duplicates). Keep the SNP with the highest score (for if I ever want to do extra filtering based on score)
  Exome_AllSNPs <- multi_full[order(multi_full[,'SNPID'],-multi_full[,'Score']),]
  Exome_AllSNPs<- Exome_AllSNPs[!duplicated(Exome_AllSNPs$SNPID),]
  
  #Filter RNASeq variants based on filtered exome variants
  RNA_VariantsInExome<- lapply(RNA_UniqueSNPID, function(x) {x[x$SNPID %in% Exome_AllSNPs$SNPID,]})

#Filter step 2: Alt/Ref coverage by PERCENTAGE OF TOTAL COVERAGE. This is in addition to the raw thresholds set when importing data
  PercentThreshold<- 0.1
  RNA_VariantsInExome_Ref<- lapply(RNA_VariantsInExome, function(x){
    x[x$RefDepth>=round(x$Coverage*PercentThreshold),]
  })
  RNA_VariantsInExome_Alt<- lapply(RNA_VariantsInExome_Ref, function(x){
    x[x$AltDepth>=round(x$Coverage*PercentThreshold),]
  })

#Recalculate median allelic expression values based on this new dataset of variants- need to re-filter by nVariants/gene
RNA_Subset_AfterExome<- lapply(RNA_VariantsInExome_Alt, function(x) {x[,c(1:16,19)]})
RNA_ReFilterAfterExome_NonCM<- lapply(RNA_Subset_AfterExome, NVariantsFilter_AllelicExpresion, NVariantsPerGene=3)


#Add columns for mean allelic expression + SD Allelic Expression
RNA_AddMean<- lapply(RNA_ReFilterAfterExome_NonCM, function(x){
  x %>% group_by(geneid) %>% mutate(MeanAllelicExpression=mean(AllelicExpression)) %>% as.data.frame
})

RNA_AddMeanAndSD<- lapply(RNA_AddMean, function(x){
  x %>% group_by(geneid) %>% mutate(SDAllelicExpression=sd(AllelicExpression)) %>% as.data.frame
})

#Merge filtered variants for writing to csv
for (i in 1:length(RNA_AddMeanAndSD)){
  RNA_AddMeanAndSD[[i]]$Sample<- RNASamples[i]
}
NonCMFilteredVariants<-ldply(RNA_AddMeanAndSD, data.frame)

#Combine all median allelic expression values for each
RNA_MedianFinal<- lapply(RNA_AddMeanAndSD, function(x) {x[!duplicated(x$geneid),]})

#bind all median allelic expression values into 1 (keepin just SNP info + Sample + MedianAllelicExpression), then write to csv
MediansAll<-  do.call(rbind, RNA_MedianFinal)
NonCM_MediansAll<- MediansAll[order(MediansAll$geneid),colnames(MediansAll) %in% c("geneid","NumberVariantsPerGene", "MedianAllelicExpression", "MeanAllelicExpression", "SDAllelicExpression", "Sample")]
```

#Export results
```{r}
#Filtered variants
write.csv(NonCMFilteredVariants, paste(writewd, "FinalFilteredVariants_PercentCoverageExome_AllNonCMVariants.csv",sep=""))

#Median allelic expression for each gene
write.csv(NonCM_MediansAll, paste(writewd, "NonCMGenes_MedianAllelicExpression_AllRNASeqVariants_CoveragePer.1_Total15_RefAlt1_3PerGene_FilteredByExomeQual40Total15RefAlt2.csv", sep=""))
```


#Test allelic expression differences for significance
Using Dunn sig test (ANNOVA post hoc)

```{r}
#Remove indels- incorrect alignment for these variants common, don't want to calculate significance using these SNPs
nChar<- apply(NonCMFilteredVariants, 2, nchar)[,7:8]
AllNonCM_NoIndels<-NonCMFilteredVariants[rowSums(nChar)==2,]

#Subset data- just need geneid, sample, and allelic expression
AllNonCM_sigTest<- AllNonCM_NoIndels[,colnames(AllNonCM_NoIndels) %in% c("geneid", "AllelicExpression", "Sample")]

#Rename Control-1/2/3/4/5 to control- will use all 5 controls as 1 "sample"
for (i in 1:5){
  sub<-paste("UK-Control-", i, sep="")
  AllNonCM_sigTest$Sample<- gsub(sub, "Control", AllNonCM_sigTest$Sample)
}

#Convert to factor
AllNonCM_sigTest$Sample<-as.factor(AllNonCM_sigTest$Sample)
AllNonCM_sigTest$Sample<- ordered(AllNonCM_sigTest$Sample, levels= c("Control",1:9,11:13,15:19))

#Split dataframe into a list of separate dfs for each gene
SplitByGene<- split(AllNonCM_sigTest, f=AllNonCM_sigTest$geneid)

#only keep gene if there's at least 1 variant in a non-control sample + 3 control samples
#At least 3 variants in a control
NControls<- unlist(lapply(SplitByGene, function(x) {
  length(grep("Control", x$Sample))
}))
EnoughControls<-SplitByGene[NControls>=3]

#At least 1 NonCM sample
detach("package:plyr")
NSamples<- lapply(EnoughControls, function(x) { x %>% filter(Sample!= "Control") %>% group_by(Sample) %>% mutate(NumberVariants=n()) %>%  filter(NumberVariants>=1) %>% as.data.frame})
EnoughSamples<- EnoughControls[unlist(lapply(NSamples,nrow))>=1]


#initialize empty list for appending significant gene/samples to
SigAllelicExpression<- data.frame()

#Run dunn test for all genes 
for (i in 1:length(EnoughSamples)){
  variants<-EnoughSamples[[i]]
  dunn<-with(variants, posthoc.kruskal.dunn.test(x=AllelicExpression, g=Sample, p.adjust.method="none"))
  pvalue<-as.data.frame(dunn$p.value[,1])
  colnames(pvalue)<- "Pvalue"
  pvalue$Sample<-rownames(pvalue)
  pvalue$gene<- variants$geneid[1]
  sig<-pvalue[pvalue$Pvalue<0.05,]
  SigAllelicExpression<- rbind(SigAllelicExpression, sig)
  
}

#Add on info about median/mean/sd of the significantly different genes, for the sig sample + across controls
SigAllelicExpression$UniqueID<- paste(SigAllelicExpression$Sample, SigAllelicExpression$gene, sep=":")
NonCM_MediansAll$UniqueID<- paste(NonCM_MediansAll$Sample, NonCM_MediansAll$geneid, sep=":")
SigGenes_Medians<- NonCM_MediansAll[NonCM_MediansAll$UniqueID %in% SigAllelicExpression$UniqueID,]


#Add in controls-mean/median/SD of allelic expression across all 5 control samples
#can't just take mean of these metrics across the 5 samples- there's different n of variants supporting these metrics
#Need to weight each variant equally
AllControls_AEMetrics<- AllNonCM_sigTest[grep("Control", AllNonCM_sigTest$Sample),]
AllControls_AEMetrics<- AllControls_AEMetrics %>% group_by(geneid) %>% mutate(Controls_MedianAllelicExpression=median(AllelicExpression),
                                                                              Controls_MeanAllelicExpression=mean(AllelicExpression),
                                                                              Controls_SDAllelicExpression= sd(AllelicExpression)) %>% as.data.frame
AllControls_AEMetrics<- AllControls_AEMetrics[!(duplicated(AllControls_AEMetrics$geneid)),c(1,4:6)]
AllControls_SigGenes<- AllControls_AEMetrics[AllControls_AEMetrics$geneid %in% SigGenes_Medians$geneid,]
                                                                      
#Merge sig gene Allelic expression values from NonCM samples with corresponding control values 
ControlsandNonCM<- full_join(SigGenes_Medians, AllControls_SigGenes, by="geneid")

#Merge this with significance info- pvalue from dunn test
Sig_AndAllelicExpressionMetrics_id<- cbind(SigAllelicExpression, ControlsandNonCM, by="UniqueID")
Sig_AndAllelicExpressionMetrics<- Sig_AndAllelicExpressionMetrics_id[,c(3,2,1,6:9,12:14)]


#Prep for export: all the SNPs that caused the significant gene/sample pairings. Will merge this with the above table into one file manually in excel.
AllFilteredSNPs<- do.call(rbind,EnoughSamples)
AllFilteredSNPs$UniqueID<- paste(AllFilteredSNPs$Sample, AllFilteredSNPs$geneid,sep=":")
AllFilteredSNPs_SigGeneSamples<- AllFilteredSNPs[AllFilteredSNPs$UniqueID %in% Sig_AndAllelicExpressionMetrics_id$UniqueID,]
```

Export results
```{r}
#significant gene/sample pairings- no info on indv SNPs
writewd<- "//Cifs2/rcdata$/UKRNASeq/VariantCalling/"
write.csv(Sig_AndAllelicExpressionMetrics, paste(writewd, "DunnSigTest_AllelicExpression_NonCMGenes_PerCoverageFilter.1.csv", sep=""))

#All SNPs that cause sig gene/sample pairings
write.csv(AllFilteredSNPs_SigGeneSamples, paste(writewd, "RNASeqVariantsInNonCMGenes_CausingSigAllelicExpression.csv", sep=""))
```

#For genes/sample pairs of interest (i.e., sig Allelic imbalance), get ALL EXOME VARIANTS

```{r}
NonCM_SigAllelicExpression<-SigAllelicExpression

#Add sample info
for (i in 1:length(NonCM_AllAllelicExpression_Exome)){
  NonCM_AllAllelicExpression_Exome[[i]]$Sample<- Samples[i]
}

#Merge all exome variants together
NonCM_ExomeVariants<- do.call(rbind, NonCM_AllAllelicExpression_Exome)
NonCM_ExomeVariants$UniqueID<- paste(NonCM_ExomeVariants$Sample, NonCM_ExomeVariants$geneid, sep=":")

#Filter exome variants- only keep variants that are found in a gene/sample pair that had sig allelic imbalance
SigAI_ExomeVariants<- NonCM_ExomeVariants[NonCM_ExomeVariants$UniqueID %in% NonCM_SigAllelicExpression$UniqueID,]
```

Export
```{r}
#All exome variants in sig gene/sample pairs
write.csv(SigAI_ExomeVariants, paste(writewd, "ExomeVariants_NonCMGenes_InGeneSamplePairSigAllelicImbalance.csv", sep=""))
```


