---
title: "R Notebook"
output: html_notebook
---

```{r}
library(MBASED)
library(dplyr)
```

functions
```{r}
summarizeASEResults_1s <- function(MBASEDOutput) {
  geneOutputDF <- data.frame(majorAlleleFrequency=assays(MBASEDOutput)$majorAlleleFrequency[,1],
                             pValueASE=assays(MBASEDOutput)$pValueASE[,1],
                             pValueHeterogeneity=assays(MBASEDOutput)$pValueHeterogeneity[,1])
  lociOutputGR <- rowRanges(metadata(MBASEDOutput)$locusSpecificResults)
  lociOutputGR$allele1IsMajor <- assays(metadata(MBASEDOutput)$locusSpecificResults)$allele1IsMajor[,1]
  lociOutputGR$MAF <- assays(metadata(MBASEDOutput)$locusSpecificResults)$MAF[,1]
  lociOutputList <- split(lociOutputGR, factor(lociOutputGR$aseID, levels=unique(lociOutputGR$aseID)))
  return(list(geneOutput=geneOutputDF,
              locusOutput=lociOutputList))
}
```


#Example from vignette
```{r}
##a quick look at the main function
args(runMBASED)

## create GRanges object for SNVs of interest.
## note: the only required column is 'aseID'
mySNVs <- GRanges(
seqnames=c('chr1', 'chr2', 'chr2', 'chr2'),
ranges=IRanges(start=c(100, 1000, 1100, 1200), width=1),
aseID=c('gene1', rep('gene2', 3)),
allele1=c('G', 'A', 'C', 'A'),
allele2=c('T', 'C', 'T', 'G')
)
names(mySNVs) <- c('gene1_SNV1', 'gene2_SNV1', 'gene2_SNV2', 'gene2_SNV3')

## create input RangedSummarizedExperiment object
mySample <- SummarizedExperiment(
  assays=list(
    lociAllele1Counts=matrix(
      c(25, 10, 22, 14),
      ncol=1,
      dimnames=list(names(mySNVs),'mySample')),
    lociAllele2Counts=matrix(c(20,16,15,16),
                             ncol=1,dimnames=list(names(mySNVs),'mySample'))),
  rowRanges=mySNVs)

##example of use
ASEresults_1s_haplotypesKnown <- runMBASED(
ASESummarizedExperiment=mySample,
 isPhased=TRUE,
 numSim=10^6,
 BPPARAM = SerialParam()
 )

summarizeASEResults_1s(ASEresults_1s_haplotypesKnown)$geneOutput
```


Using our data
```{r}
#Read in vcf
file<- "D:/UKy/RNASeq_RawVcf/CMGenesROI_AllelicDepthFromVCF/AllelicDepthFinal_GOIFilteredvcf2.vcf.gz.txt"

#Process, get allelic depth
mySNVs<-AllelicExpressionFromVCF(file, GOIRanges, 15,1,1,3,0 ) 

#subset and rename
mySNVs<-mySNVs[,c(1,2,3,7,8,11,12,13)]
colnames(mySNVs)<-c("seqnames", "start", "end", "Ref", "Alt", "allele1", "allele2", "aseID")

#Add unique column for each snp- geneid_Number
mySNVs<-mySNVs %>% group_by(aseID) %>% mutate(ID=paste(aseID, 1:n(),sep="_")) %>% as.data.frame

#Convert to granges
mySNVs_gr<- makeGRangesFromDataFrame(mySNVs, keep.extra.columns = TRUE)
names(mySNVs_gr)<-mySNVs_gr$ID

#Create summarized experiment with info on allelic depth
myOwnSample<-SummarizedExperiment(
  assays=list(
    lociAllele1Counts=matrix(
      mySNVs_gr$allele1,
      ncol=1,
      dimnames=list(names(mySNVs_gr),'mySample')),
    lociAllele2Counts=matrix(
      mySNVs_gr$allele2,
      ncol=1,
      dimnames=list(names(mySNVs_gr), "mySample"))),
  rowRanges=mySNVs_gr
    )

##Get results
results<-as.data.frame(assays(ASEresults_1s_haplotypesKnown)$pValueASE)

length(results[results$mySample<0.05,]) #37 out of 42 genes are sig

```

