SAMPLES=['1', '2', '3', '4', '5', '6', '7', '8', '9', '11', '12', '13', '15','16','17','18', '19']

#Outputs desired
rule all:
        input:
                expand('vcfs/{sample}_Merged.vcf.idx', sample=SAMPLES),
		expand('vcfs/{sample}_Merged_Selected.vcf', sample=SAMPLES),
		expand('{sample}_output.table', sample=SAMPLES)

#Create indexed vcf, if not already created. Required to exist in same directory as vcf file for later steps
rule create_vcf_index:
        input:
                "vcfs/{sample}_Merged.vcf"
        output:
                "vcfs/{sample}_Merged.vcf.idx"
        shell:
                "gatk IndexFeatureFile --input {input}"

#Filter vcfs to reduce compute time- only include biallelic SNPs
rule filter_vcf:
        input:
                vcf="vcfs/{sample}_Merged.vcf",
                genome="/mnt/d/UKy/hg19.fa"
        output:
                "vcfs/{sample}_Merged_Selected.vcf"
        shell:
                "gatk SelectVariants -R {input.genome} -V {input.vcf} -O {output} --select-type-to-include SNP --restrict-alleles-to BIALLELIC"

#Run ASEReadCounter on filtered SNPs + raw bams
rule run_ASEReadCounter:
        input:
                bam="/mnt/d/UKy/Merged_Bams/{sample}_Merged.alignments.bam",
                vcf="vcfs/{sample}_Merged_Selected.vcf"
        output:
                "{sample}_output.table"
        shell:
                "gatk ASEReadCounter -R /mnt/d/UKy/hg19.fa -I {input.bam} -V {input.vcf} -O {output}"
