---
title: "R Notebook"
output: html_notebook
---

#Analysis for filtering junctions to identify aberrant splice sites

Functions from "AberrantSplicing_Functions.Rmd" used here

Overall:

1. Read in CM and controls junctions. Each sample should be a separate csv file, optionally filtered by GOI, with at least the columns:
    seqnames (chromosome #)
    start (start of junction)
    end (end of junction)
    Score (Depth of junction)
    Name (chr#:start-end)
    geneid (gene junction is found in)

2. Create filtered controls lists, used for filtering CM junctions:
    -Control junctions
    -Control splice sites (=Canonical Splice Sites)

3. Filter CM junctions by:
    -Depth
    -Presence in Control junctions
    -presence in canonical splice sites

4. Annotate junctions with Norm score (used to priortize junctions for downstream analysis, e.g., IGV)
    

```{r}
library(dplyr)
```

#All settings used for filters:
```{r}
#Creating filtered controls list
CountThreshold<-5 #makes a large difference in total n of filtered junctions
NControlsThreshold<-3 #makes virtually no difference

#Creating lists of canonical splice sites:
DepthPercentageFilter <- 0.5 #makes a large difference in total n of filtered junctions

#Filter CM junctions by depth (percentage of mean)
Percentage<-0.1 #makes a large difference in total n of filtered junctions

#Filter CM junctions by "uniqueness"- how many times can it be found in all CM samples?
nThreshold=1
```


#1. Read in data- junctions for CM samples and Control samples all in separate csv files (bed file format)
Previously filtered to only include junctions found in genes of interest
```{r}
#Set working directory
wd<- "D:/UKy/MergingRNASeq/MergedRuns_junctions.bed/"

#Set list of samples needed- CM samples numbered as 1-19, Controls numbered as UK-Control-1-5
Samples<-c(1:19)
FileList<- paste(wd, "GOIFilteredJunctions_Sample", Samples, ".csv", sep="")
ControlFileList<- paste(wd,"GOIFilterJunctions_","UK-Control-", c(1:5), ".csv", sep="")

#Read in all junction files (bed files)
AllCMJunctions<- lapply(FileList, read.csv, header=TRUE, stringsAsFactors=FALSE)
AllControlJunctions<- lapply(ControlFileList, read.csv, header=TRUE, stringsAsFactors=FALSE)

#reformat- remove first column from each file
AllCMJunctions<- lapply(AllCMJunctions, function(x) {x[,-1]})
AllControlJunctions<- lapply(AllControlJunctions, function(x) {x[,-1]})

#Subset to only the columns necessary- seqnames, start, end, Score, geneid
ColumnsNeeded<-c("seqnames", "start", "end", "Score", "geneid", "Name")
AllCMJunctions<-lapply(AllCMJunctions,function(x){x[,colnames(x) %in% ColumnsNeeded]})
AllControlJunctions<-lapply(AllControlJunctions,function(x){x[,colnames(x) %in% ColumnsNeeded]})

head(AllCMJunctions[[1]])

paste(length(unique(unlist(lapply(AllControlJunctions, function(x){x[,5]})))), "total unique control junctions")
```

#Adjust depth of junctions based on depth of sequencing- total #reads mapped to gene
First, need to remove any samples from analysis with very low counts (<10million total mapped reads)
```{r}
#Read in raw count data
readwd<-"C:/Users/jenna/OneDrive/Documents/UKy_RNASeq/RawCounts/"
countsInput<- read.csv(paste(readwd, "RawCounts_Run1S10.14_Run2Controls_AllOthersMerged.csv", sep =""),row.names = "GeneID")
countsInput<-countsInput[,-1]

CM_TotalCounts<-colSums(countsInput[,1:19])
Controls_TotalCounts<-colSums(countsInput[20:24])

#Remove samples with low counts
AllCMJunctions_LowSampleRemove<- AllCMJunctions[CM_TotalCounts>10000000] #Removes 2 samples
AllControlJunctions_LowSampleRemove<-AllControlJunctions[Controls_TotalCounts>1000000] #Removes no samples

#Get factors used to adjust depth- total counts/min(total counts across all samples)
All_TotalCounts<-colSums(countsInput[,colSums(countsInput)>10000000])
CM_DepthAdjust<-All_TotalCounts[1:17]/min(All_TotalCounts)
Controls_DepthAdjust<-All_TotalCounts[18:22]/min(All_TotalCounts)

#Run adjust depth function for controls and CM
AllCMJunctions_DepthAdjusted<-AdjustDepth(AllCMJunctions_LowSampleRemove,CM_DepthAdjust)
AllControlJunctions_DepthAdjusted<-AdjustDepth(AllControlJunctions_LowSampleRemove,Controls_DepthAdjust)
```


#2. Create filtered controls lists, used for filtering CM junctions:

#Create list of Control Junctions (default settings)
```{r}
FilteredControlJunctions<-CreateControlJunctionsList(AllControlJunctions_DepthAdjusted, CountThreshold=CountThreshold, NControlsThreshold=NControlsThreshold)
```

#Create list of Canonical Splice sites(Default settings)
```{r}
CanSpliceSites<-CreateCanonicalSpliceSiteList(AllControlJunctions_DepthAdjusted, DepthPercentageFilter = DepthPercentageFilter)
CanASpliceSites<-CanSpliceSites[[1]]
CanBSpliceSites<-CanSpliceSites[[2]]
```

#3. Filter CM Junctions (3 filters total)

1) Remove junction based on depth of reads supporting the junction. Two options (can both be used if desired):  
  a) Remove if depth is below some static threshold (default=10)  
  b) Remove if depth is below some percentage (default=10%) of the mean depth of all junctions in that sample.  
    Example: Junction chr1_100-200 has a depth of 10 in Sample 1 in gene X. 
            On average, all junctions in gene X in Sample 1 have a depth of 200. 
            Since Junction chr1_100-200 is at <10% mean depth, this junction is removed in this filter
    
2) Remove junction if found in list of Control Junctions

3) Remove junction if BOTH or NEITHER splice sites of the junction are classified as "Canonical Splice Sites".  
  Example: CM junction chr1_100-300 is split into SpliceSiteA_chr1_100 and SpliceSiteB_chr1_300. 
          SpliceSiteA_chr1_100 is considered "Canonical" (i.e., found in the "Canonical Splice Sites" list constructed previously). 
          SpliceSiteB_chr1_300 is not considered Canonical. This junction, therefore,  will not be removed in this filter.

#Remove junction based on depth of reads supporting the junction- using Percentage based (option b; default (10%)) method here
```{r}
CMJunctions_DepthFilter<-lapply(AllCMJunctions_DepthAdjusted, MeanDepthFilter, Percentage=Percentage)

print(paste(mean(unlist(lapply(AllCMJunctions_DepthAdjusted, nrow))), "splice sites per sample, on average before"))
print(paste(mean(unlist(lapply(CMJunctions_DepthFilter, nrow))), "splice sites per sample, on average after"))
```

#Remove junction if found in list of Control Junctions
```{r}
CMJunctions_ControlJunctionsFilter<- lapply(CMJunctions_DepthFilter, ControlJunctionsFilter, ControlJunctions=FilteredControlJunctions)

print(paste(mean(unlist(lapply(CMJunctions_DepthFilter, nrow))), "splice sites per sample, on average before"))
print(paste(mean(unlist(lapply(CMJunctions_ControlJunctionsFilter, nrow))), "splice sites per sample, on average after"))
```

#Remove junction if found in more than X CM samples (including the one being tested, default 4)
```{r}
CMJunctions_CommonCMRemoved<-CMJunctionsFilter(CMJunctions_ControlJunctionsFilter,nThreshold )

print(paste(mean(unlist(lapply(CMJunctions_ControlJunctionsFilter, nrow))), "splice sites per sample, on average before"))
print(paste(mean(unlist(lapply(CMJunctions_CommonCMRemoved, nrow))), "splice sites per sample, on average after"))
```

#Remove junction if BOTH or NEITHER splice sites of the junction are classified as "Canonical Splice Sites".  
```{r}
#Add splice site IDs to CM samples
CMSplitJunctions<- lapply(CMJunctions_CommonCMRemoved, SpliceSiteSplit)

#Add annotation detailing whether each side of the junction is "canonical" and what the mean depth for that canonical junction is (in controls)
CMCanonicalAnnotatedJunctions<- lapply(CMSplitJunctions, FilterOutBothNeitherCanonical, CanASpliceSites,CanBSpliceSites)

#Select just junctions where exactly one side is canonical
AllOneCanonicalJunctions<- lapply(CMCanonicalAnnotatedJunctions, function(x) {x[x$OneCanonical=="TRUE",]})

print(paste(mean(unlist(lapply(CMCanonicalAnnotatedJunctions, nrow))), "splice sites per sample, on average before"))
print(paste(mean(unlist(lapply(AllOneCanonicalJunctions, nrow))), "splice sites per sample, on average after"))
```

#4. Annotate junctions with Norm score (used to priortize junctions for downstream analysis, e.g., IGV)
```{r}
CMJunctions_NormScoreAnnotation<-lapply(AllOneCanonicalJunctions, GetNormScore)


print(paste(mean(unlist(lapply(CMJunctions_NormScoreAnnotation, nrow))), "splice sites per sample, on average"))
print(paste(mean(unlist(lapply(CMJunctions_NormScoreAnnotation, function(x) {nrow(x[x$NormScore>=0.1,])}))), "splice sites per sample, on average, with 10% norm score"))
```


#Merge results for all samples together, use for exporting
```{r}
#Remove any samples that have no filtered junctions
failingSamples<-c()

count=1
for (i in 1:length(CMJunctions_NormScoreAnnotation)){
  df<-CMJunctions_NormScoreAnnotation[[i]]
  if(nrow(df) ==0){
    failingSamples[count]=Samples[i]
    count=count+1
  }
}

passingSamples<-Samples[!(Samples %in% failingSamples)]
passingJunctions<-CMJunctions_NormScoreAnnotation[!(1:length(CMJunctions_NormScoreAnnotation) %in% failingSamples)]

#Add sample information to each junction
for (i in 1:length(passingJunctions)){
  passingJunctions[[i]]$SampleID<-passingSamples[i]
}

AllFilteredJunctions<- do.call(rbind,passingJunctions)

paste(nrow(AllFilteredJunctions), "filtered junctions")
paste(nrow(AllFilteredJunctions[AllFilteredJunctions$NormScore>0.1,]), "junctions in with Norm score>10%")

paste(sum(AllFilteredJunctions[AllFilteredJunctions$geneid=="7273",20] %in% c(2,3,4,9)), "TTN junctions in S2/3/4/9")

paste(sum(!(AllFilteredJunctions[AllFilteredJunctions$geneid=="7273",20] %in% c(2,3,4,9))), "TTN junctions NOT in S2/3/4/9")

AllFilteredJunctions[AllFilteredJunctions$geneid=="7273",]
```

```{r}
settings<-paste("Settings_",CountThreshold, NControlsThreshold, DepthPercentageFilter, Percentage, nThreshold,sep="_")

wd<-"~/UKy/"
write.csv(AllFilteredJunctions,paste(wd, "AllFilteredJunctions_", settings, ".csv",sep="_"))

```


Considering this final list of filtered junctions: is there an exome variant nearby (say, within 20bp) of the "interesting" side of the junction?
```{r}
#Read in exome variants
ExomeSNPs<-read.csv( "C:/Users/jenna/OneDrive/Documents/UKy/AllExomeVariants_NoFilter.csv",row.names="X")

FindSNPsNearJunction<-function(Junctions, SNPs, InterestingAorB,DistanceFromJunction){
  #remove chr1 from splice sitea/b names
  Junctions$SpliceSiteA<- as.integer(gsub(".*:", "", Junctions$SpliceSiteA))
  Junctions$SpliceSiteB<- as.integer(gsub(".*:", "", Junctions$SpliceSiteB))
  
  #Depending on whether you're looking at non-canonical A or B sites, get the start/end of the junction
  if (InterestingAorB=="A"){
    Sites<-Junctions[Junctions$A_Canonical=="FALSE", colnames(Junctions) %in% c("seqnames", "SpliceSiteA",
                                                                                "geneid", "NormScore", "SampleID")]
  } else{
    Sites<-Junctions[Junctions$B_Canonical=="FALSE", colnames(Junctions) %in% c("seqnames", "SpliceSiteB",
                                                                                "geneid", "NormScore", "SampleID")]
  }
  colnames(Sites)<-c("Site", "seqnames", "geneID","NormScore","Sample")
  Sites$Site<-as.integer(Sites$Site)

 
 #Add in a range of nt to the start/end of the site
 Sites$start<-Sites$Site-DistanceFromJunction
 Sites$end<-Sites$Site+DistanceFromJunction

 
 #Remove sample 4 (for now)- no exome data rn
 Sites<-Sites[!(Sites$Sample=="4"),]
 
 #Remove any samples from the SNP data that aren't in the Asites
 SNPs_Sites<-SNPs[SNPs$Sample %in% Sites$Sample,]
 
 #Split exome into 1 df per sample
 SNPs_Sites<-split(SNPs_Sites, f=SNPs_Sites$Sample)
 
 #Convert SNPs to granges
 SNPs_Sites_gr<-lapply(SNPs_Sites, function(x) {makeGRangesFromDataFrame(x[,c(1:3,7:8,11:12,13,19)],
                                                                                     keep.extra.columns = TRUE)})
 
 #Split junctions into 1 df per sample
 Sites<-split(Sites, f=Sites$Sample)
 
 #convert junctions to granges object
 Sites_gr<- lapply(Sites, function(x) {makeGRangesFromDataFrame(x, keep.extra.columns = TRUE)})
 
 #Get overlaps between junction range and exome variants, for each sample
 AllJunctionsInRange<-list()
 for (i in 1:length(Sites)){
  site<-Sites_gr[[i]]
  snps<-SNPs_Sites_gr[[i]]
  JunctionsInRange<-subsetByOverlaps(site, snps)
  AllJunctionsInRange[[i]]<-JunctionsInRangedi
}
 
 return(AllJunctionsInRange)
}





test<-FindSNPsNearJunction(AllFilteredJunctions, ExomeSNPs, "A", 50)

jTest<-AllFilteredJunctions[AllFilteredJunctions$geneid=="7273",]
sTest<-ExomeSNPs[ExomeSNPs$Sample==2 & ExomeSNPs$geneid=="7273",]













```









```{r}
#sessionInfo()
```

