---
title: "Pathogenic Splice Site Identification"
output: html_notebook
---
Import libraries
```{r}
suppressPackageStartupMessages(library(Homo.sapiens))
suppressPackageStartupMessages(library(tidyverse)) #Important: load this library, NOT dplyr, etc separately!!!
suppressPackageStartupMessages(library(org.Hs.eg.db))
```


Functions
```{r}
#Convert bed--> genomic ranges --> add gene IDs --> select gene ID of interest
GetJunctionGeneID<- function(bed, genes, GeneIDs) {
  #Set column names
  colnames(bed)<- c("Chrom", "Start", "End", "Name", "Score", "Strand", "ThickStart", "ThickEnd","ItemRGB", "BlockCount", "BlockSizes", "BlockStarts")
  #colnames(bed)<- c("Chrom", "Start", "End", "Name", "Score", "Strand", "ThickStart", "ThickEnd","ItemRGB", "BlockCount", "BlockSizes", "BlockStarts", "Sample")
  #Convert bed file to genomic ranges, add score as metadata column
  GR<- makeGRangesFromDataFrame(bed)
  values(GR)<- DataFrame(Score=bed$Score, Name=bed$Name)
  #values(GR)<- DataFrame(Score=bed$Score, Name=bed$Name, Sample=bed$Sample)
  
  #Get overlaps of GRanges bed info and all known genes
  ranges<- subsetByOverlaps(GR,genes)
  
  #Get the geneID names, associate back with the overlap GRanges object
  hits<- findOverlaps(GR,genes)
  geneid <- CharacterList(split(genes$gene_id[subjectHits(hits)],queryHits(hits)))
  mcols(ranges) <- DataFrame(mcols(ranges), geneid)
  
  Junctions_GeneIDs<- as.data.frame(ranges)
  GeneJunctions_GeneIDs<- Junctions_GeneIDs[Junctions_GeneIDs$geneid %in% GeneIDs,]
  return(GeneJunctions_GeneIDs)
}

#Filter junctions based on pre-built control list of junctions + depth
FilterJunctions<- function(ControlJunctions, SampleJunctions, DepthFilter){
  #Get rid of any in control list
  ControlJunctionsInSample<- SampleJunctions$Name %in% ControlJunctions$Name
  UniqueTTNJunctions<- SampleJunctions[!ControlJunctionsInSample,]
  
  #Get rid of any with fewer reads supporting it than the depth filter
  UniqueSample_Depth<- UniqueTTNJunctions[UniqueTTNJunctions$Score>DepthFilter,]
  return(UniqueSample_Depth)
}

#Alternative to above function- rather than setting a depth filter, filter junctions based on the mean depth of junctions of that gene of that sample. Also removes any in control list.
FilterJunctions_BasedOnMeanDepth<- function(ControlJunctions, SampleJunctions, Percentage){
  SampleJunctions$geneid<- as.character(SampleJunctions$geneid)
  #Get rid of any with fewer reads supporting it than 10% of the mean number of reads supporting junctions 
  #of THAT GENE in THAT SAMPLE
 SampleJunctions_Depth<-as.data.frame(group_by(SampleJunctions, geneid) %>% mutate(mean=mean(Score))) %>% filter(Score>(mean*Percentage))
  
  #Get rid of any in control list
  ControlJunctionsInSample<- SampleJunctions_Depth$Name %in% ControlJunctions$Name
  UniqueTTNJunctions<- SampleJunctions_Depth[!ControlJunctionsInSample,]
}

#Function that splits "name" into two columns: splice site A and B, and concatanates with chromosome to give unique splice site
SpliceSiteSplit<- function(Junctions){
  AllJunctions<- unlist(lapply(strsplit(Junctions$Name, ":"), function(x){x[2]}))
  SpliceSites<- strsplit(AllJunctions, "-")
  SpliceSiteA<- unlist(lapply(SpliceSites, function(x){x[1]}))
  SpliceSiteB<- unlist(lapply(SpliceSites, function(x){x[2]}))
  Junctions$SpliceSiteA<- paste(Junctions$seqnames, SpliceSiteA,sep=":")
  Junctions$SpliceSiteB<- paste(Junctions$seqnames, SpliceSiteB,sep=":")
  return(Junctions)
}

#For determining canonoical sites: Get rid of any with fewer reads supporting it than some percentage of the mean number of reads supporting junctions of THAT GENE in THAT SAMPLE
FilterControlJunctions_BasedOnMeanDepth<- function(Junctions, Percentage){
  Junctions$geneid<- as.character(Junctions$geneid)
 SampleJunctions_Depth<-as.data.frame(group_by(Junctions, geneid) %>% mutate(mean=mean(Score))) %>% filter(Score>(mean*Percentage))
}

#For determining canonical splice sites: group by splice site A or B and sum all of A or all of B- i.e., if an A splice site has 2 B splice sites, sum the scores for both sites for the A and keep separate for the B
GetCanonicalSpliceSiteSum<-function(DepthFilteredJunctions, AorB){
  SpliceSite<-paste("SpliceSite", AorB, sep="")
  SummedSpliceA<- DepthFilteredJunctions %>% group_by_at(vars(SpliceSite))%>% summarize(MeanScore=sum(Score)) %>% as.data.frame
}

#For adding column to partially filtered CM junctions annotating whether each side of the junction is "canonical" and what the mean depth of that "canonical" junction is in controls
FilterOutBothNeitherCanonical<- function(SplitJunction, MergedAControlSpliceSites, MergedBControlSpliceSites){
  #Splice site A not in list:
  SplitJunction$A_Canonical<- SplitJunction$SpliceSiteA %in% MergedAControlSpliceSites$SpliceSiteA
 
  #Splice site B not in list:
  SplitJunction$B_Canonical<-SplitJunction$SpliceSiteB %in% MergedBControlSpliceSites$SpliceSiteB

  #Neither splice site canonical
  SplitJunction$Sum<- rowSums(SplitJunction[,11:12])
  SplitJunction$NeitherCanonical<- SplitJunction$Sum==0
  
  #Both splice sites canonical
  SplitJunction$BothCanonical<- SplitJunction$Sum==2
  
  #One splice site canonical one not
  SplitJunction$OneCanonical<- SplitJunction$Sum==1
  
  #Depth of junction for controls
  ControlDepths<- merge(SplitJunction, MergedAControlSpliceSites, by="SpliceSiteA",all.x=TRUE)
  ControlDepths<-merge(ControlDepths, MergedBControlSpliceSites,by="SpliceSiteB",all.x=TRUE)
  ControlDepths$MeanControlA<- rowMeans(ControlDepths[,17:21])
  ControlDepths$MeanControlB<- rowMeans(ControlDepths[,22:26])
  ControlDepths[is.na(ControlDepths)]<- 0
  return(ControlDepths[,c(3:7,9,10,2,1,8,11,12,14:16,27,28)])
  
}

```

#Import data

For running analysis on merged data set: Create merged dataset#Import raw junctions for individual samples (difference sources, described below) and merge into one dataset

**Skip this section after performing once
```{r}
#Three types of data: 
  #1: samples run in both run 1 and run 2, fastqs merged --> alignment
  #2: samples run in only run 1 (10 and 14)
  #3: samples run in only run 2 (controls)

extension<- ".junctions.bed"

############
#Dataset 1 #
############

#set list of files- merged files
merged_readwd<- "E:/UKy/MergingRNASeq/MergedRuns_junctions.bed/"
merged_samples<- c(1:9,11:13,15:19)
merged_files<-c(paste(merged_samples, "_Merged", extension,sep=""))

#Read in junctions data
merged_Alljunctions<-list()
for (i in 1:length(merged_files)){
  file<- read.table(paste(merged_readwd, merged_files[i], sep=""))
  #file$Sample<- merged_samples[i]
  merged_Alljunctions[[i]]<- file
}

############
#Dataset 2 #
############

#set list of files- run 1 files (10 and 14)
run1_readwd<- "//Cifs2/rcdata$/UKRNASeq/RawFiles/Run1-highrRNA/"
run1_samples<- c(10,14)
run1_files<-c(paste(run1_samples, extension,sep=""))

#Read in junctions data
run1_Alljunctions<-list()
for (i in 1:length(run1_files)){
  file<- read.table(paste(run1_readwd, run1_files[i], sep=""))
  #file$Sample<-run1_samples[i]
  run1_Alljunctions[[i]]<- file
}

############
#Dataset 3 #
############

#set list of files- run 2 files (controls)
run2_readwd<- "//Cifs2/rcdata$/UKRNASeq/RawFiles/Bed/"
run2_samples<- paste("UK-Control-", c(1:5), sep="")
run2_files<-c(paste(run2_samples, extension,sep=""))

#Read in junctions data
run2_Alljunctions<-list()
for (i in 1:length(run2_files)){
  file<- read.table(paste(run2_readwd, run2_files[i], sep=""))
  #file$Sample<- run2_samples[i]
  run2_Alljunctions[[i]]<- file
}



#Merge all CM junctions together into one list (Samples 14 and 10 in run 1, all others merged) ), then into one DF
#Keep control junctions separate
AllBed<- c(merged_Alljunctions, run1_Alljunctions)
#CMjunctions<- do.call(rbind, AllCMjunctions)
AllControlBed<- run2_Alljunctions
```


#Get junctions in genes of interest (in both CM samples and controls)
**Skip this section after performing once
```{r}
#Read in txdb- only needs to be done once
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
  
#Get full list of genes and genes of interest
genes <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)

#Input genes of interest- as gene name
geneNames<- read.csv("//Cifs2/rcdata$/UKRNASeq/PathogenicSpliceSiteAnalysis/GenesofInterest.csv",header=FALSE)

#Convert to entrez gene ids
GOI<- unname(mapIds(org.Hs.eg.db, keys=as.character(geneNames$V1), column="ENTREZID",keytype="SYMBOL"))

#Get Genes of interest junctions, in CM samples
AllGOIJunctions<- lapply(AllBed, GetJunctionGeneID, genes=genes, GeneIDs=GOI)

#Get Genes of interest junctions, in controls 
AllGOIControlJunctions<- lapply(AllControlBed, GetJunctionGeneID, genes=genes, GeneIDs=GOI)

#Write all GOI filtered junctions to csv
  CMSamples<-c(merged_samples, run1_samples)
  for (i in 1:length(CMSamples)){
          write.csv(AllGOIJunctions[[i]], 
                    paste(merged_readwd, "GOIFilteredJunctions_Sample", CMSamples[i],".csv", sep="") )
  }
  
  for (i in 1:length(run2_samples)){
          write.csv(AllGOIControlJunctions[[i]], 
                    paste(merged_readwd, "GOIFilterJunctions_", run2_samples[i], ".csv", sep=""))
  }
```


#Read in bed files: created in first two chunks (GOI filtering)
**Start here after first two chunks have been performed once

```{r}
#No longer in use
#For run2 only alignment
#wd<- "//Cifs2/rcdata$/UKRNASeq/RawFiles/Bed/"

#Samples<- c(1:3,5:9, 11:13,15:19)
#FileList<- paste(wd, Samples, ".junctions.bed", sep="")
#AllBed<- lapply(FileList, read.csv, header=TRUE, stringsAsFactors=FALSE)
#ControlFileList<- paste("UK-Control-", 1:5, ".junctions.bed", sep="")
#AllControlBed<- lapply(ControlFileList, read.csv, header=TRUE, stringsAsFactors=FALSE)

#**perform GOI filtering before continuing
```

```{r}
#For merged alignment
wd<- "E:/UKy/MergingRNASeq/MergedRuns_junctions.bed/"

#get list of samples needed
Samples<-c(1:19)
FileList<- paste(wd, "GOIFilteredJunctions_Sample", Samples, ".csv", sep="")
ControlFileList<- paste(wd,"GOIFilterJunctions_","UK-Control-", c(1:5), ".csv", sep="")

#Read in bed files
AllGOIJunctions<- lapply(FileList, read.csv, header=TRUE, stringsAsFactors=FALSE)
AllControlBed<- lapply(ControlFileList, read.csv, header=TRUE, stringsAsFactors=FALSE)

#reformat
AllGOIJunctions<- lapply(AllGOIJunctions, function(x) {x[,-1]})
AllControlBed<- lapply(AllControlBed, function(x) {x[,-1]})
```


#Overall filtering performed

Prefilter: GOI

Remove from "interesting" list if:
1) Junction found in > 1 control
2) Junction supported by >10 reads in any 1 control
3) Junction isn't supported by at least 5 reads in that sample 
4) Both sides of junction are canonical (found in Side A and Side B list)
5) Both sides of junction are non-canonical
6) Filter based on "norm" score: depth of junction in CM sample / mean depth of CANONICAL side of that junction in control samples



#Filter #1/2: Remove junctions from CM samples based on junction presence in controls

Create list of "control junctions"
```{r}
#Convert list of all junctions found in controls --> to data.frame with 5 columns (1 for each control sample) for each junction. 

#Rename columns to include sample info
AllControlBed_JustScores<- list()
for (i in 1:length(AllControlBed)){
       df<- AllControlBed[[i]]
       df<- df[,c(7,6)]
       colnames(df)<- c("Name", paste("Control", i, "_Score", sep=""))
       AllControlBed_JustScores[[i]]<- df
}

#Merge into one df
AllGOIControlJunctions_Merged<-Reduce(function(...) merge(..., by = c('Name'), all=TRUE), AllControlBed_JustScores)

#Convert NA to 0
AllGOIControlJunctions_Merged[is.na(AllGOIControlJunctions_Merged)]<- 0

#Add column with maximum score for each junction
AllGOIControlJunctions_Merged$Max<-apply(AllGOIControlJunctions_Merged[,2:6], 1, max)

#Add column with Number of samples that have that junction
AllGOIControlJunctions_Merged$SamplesWithJunction<- rowSums(AllGOIControlJunctions_Merged[,2:6]>0)

#Get list of control junctions to REMOVE from CM junctions: Appears in >1 control samples + supported by > 5 reads
ControlJunctions_Filtered<- AllGOIControlJunctions_Merged[AllGOIControlJunctions_Merged$Max>5,]
ControlJunctions_Filtered<- ControlJunctions_Filtered[ControlJunctions_Filtered$SamplesWithJunction>1,]
```


Filter CM junctions based on filtered Control junctions
```{r}
#Filter by removing control junctions
AllFilteredJunctions<- lapply(AllGOIJunctions, function(x) {x[!(x$Name %in% ControlJunctions_Filtered$Name),]})

sum(unlist(lapply(AllFilteredJunctions, nrow))) #9,906 filtered junctions
sum(unlist(lapply(AllGOIJunctions, nrow))) #40,076 total junctions
```


#Filter #3: Depth threshold
```{r}
DepthThreshold<- 10
AllFilteredJunctions<- lapply(AllFilteredJunctions, function(x) {x[x$Score>DepthThreshold,]}) 
sum(unlist(lapply(AllFilteredJunctions,nrow)))#2,226 filtered junctions
```


#Filter #4: Remove in both sides of CM junction are "canonical"

Canonical defined as occuring in ALL control samples with at least 50% (this may change, variable in function that can be changed) of mean depth for that gene for that sample

These canonical splice sites for all controls are merged into 1 df for "A" splice sites and 1 df for "B" splice sites

Creating canonical splice site lists:
```{r}
#Using unfiltered control junctions- only filtered by GOI
AllGOIControlJunctions<- AllControlBed

#Split splice sites into separate, unique column
AllSplitControlJunctions<- lapply(AllGOIControlJunctions, SpliceSiteSplit)

#Filter all control junctions by depth
AllDepthFilteredControlJunctions<- lapply(AllSplitControlJunctions, FilterControlJunctions_BasedOnMeanDepth, Percentage=.80)

#Get all A and B spice sites (summed) for each sample:
#i.e., if an A splice site has 2 B splice sites, sum the scores for both sites for the A. Repeat for B splice sites.
AllAControlSpliceSites<- lapply(AllDepthFilteredControlJunctions, function(x) {
        x %>% group_by(SpliceSiteA) %>% summarize(MeanScore=sum(Score)) %>% as.data.frame
})

AllBControlSpliceSites<- lapply(AllDepthFilteredControlJunctions, function(x) {
        x %>% group_by(SpliceSiteB) %>% summarize(MeanScore=sum(Score)) %>% as.data.frame
})


#Rename columns to include sample info
for (i in 1:length(AllAControlSpliceSites)){
  colnames(AllAControlSpliceSites[[i]])<- c("SpliceSiteA", paste(i, "MeanScore", sep=":"))
}

for (i in 1:length(AllBControlSpliceSites)){
  colnames(AllBControlSpliceSites[[i]])<- c("SpliceSiteB", paste(i, "MeanScore", sep=":"))
}

#Merge all A splice sites into one df, all B into one df
#Only keeps splice sites present in all 5 samples
MergedAControlSpliceSites<- AllAControlSpliceSites %>% Reduce(function(dtf1, dtf2) left_join(dtf1, dtf2, by="SpliceSiteA"), .) %>% na.omit
MergedBControlSpliceSites<- AllBControlSpliceSites %>% Reduce(function(dtf1, dtf2) left_join(dtf1, dtf2, by="SpliceSiteB"), .) %>% na.omit
```

```{r}
nrow (MergedAControlSpliceSites) + nrow (MergedBControlSpliceSites) #2124 canonical junctions if 50% (1818 if 80%)- present in ALL CONTROLS
```

Filter CM samples based on canonical junctions (MergedAControlSpliceSites and MergedBControlSpliceSites)
```{r}
#Remove Sample 10 
AllFilteredJunctions<- AllFilteredJunctions[c(1:9,11:19)]

#Add splice site IDs to CM samples
AllSplitJunctions<- lapply(AllFilteredJunctions, SpliceSiteSplit)

#Add annotation detailing whether each side of the junction is "canonical" and what the mean depth for that canonical junction is (in controls)
AllCanonicalAnnotatedJunctions<- lapply(AllSplitJunctions, FilterOutBothNeitherCanonical, MergedAControlSpliceSites,MergedBControlSpliceSites)

#Select just junctions where exactly one side is canonical
AllOneCanonicalJunctions<- lapply(AllCanonicalAnnotatedJunctions, function(x) {x[x$OneCanonical=="TRUE",]})

sum(unlist(lapply(AllOneCanonicalJunctions, nrow))) #1276 junctions (50% in canonical definition) 1104 (80% in canonical definition)

```

#Filter 6: Add norm score
```{r}
NormScore<- lapply(AllOneCanonicalJunctions, function(x){
  x$MeanBothSites<- rowSums(x[,16:17])
  x$NormScore<- x$Score/x$MeanBothSites
  return(x)
})
```

Add in column for sample number, merge together
```{r}
Samples<-Samples[c(1:9,11:19)]
for (i in 1:length(NormScore)){
  NormScore[[i]]$SampleID<- Samples[i]
}

multi_full<- do.call(rbind,NormScore)

nrow(multi_full[multi_full$NormScore>0.1,]) #176 if 50% for canonical def, 99 if 80%
```

```{r}
write.csv(multi_full, paste(wd, "MERGED_AllFilteredCMJunctions_ControlRemoved5_Depth10_1CanSite_Can.8_Full62GeneList.csv"))
```



#For optimizing settings of filters: function to run entire analysis, output whether or not positive control junctions are detected

```{r}
#Set positive controls we want to discover in CM junctions
PositiveControlJunctions<- c("chr2:179584992-179585265",
                             "chr2:179603079-179603867",
                             "chr2:179603167-179603867",
                             "chr2:179603229-179603867")

test[test$Name %in% PositiveControlJunctions,]



#Get Genes of interest junctions, in CM samples
AllGOIJunctions<- lapply(AllBed, GetJunctionGeneID, genes=genes, GeneIDs=GOI)
  
#Get junctions in genes of interest in controls
AllGOIControlJunctions<- lapply(AllControlBed, GetJunctionGeneID, genes=genes, GeneIDs=GOI)
c(2:6,8:15,17)

test<-EntireFiltering(c(2:17),AllGOIJunctions, AllGOIControlJunctions, 
                      ControlsListNReads=5, 
                      RawDepthFilter=10, 
                      ControlSiteFilteringPercentage=.8,
                      PositiveControlJunctions)

EntireFiltering<- function(SamplesToUse, AllGOIJunctions, AllGOIControlJunctions, ControlsListNReads,RawDepthFilter,ControlSiteFilteringPercentage, PositiveControlJunctions){
    #Print out input settings
    print("INPUT SETTINGS")
    print(paste("Number of reads threshold for considering JUNCTION as control (Filter 2):",ControlsListNReads))
    print(paste("Raw Depth Filter (Filter 3):", RawDepthFilter))
    print(paste("Percent of Mean reads in gene for that sample for considering SPLICE SITE as canonical (Filter 4):", ControlSiteFilteringPercentage))
    print("")
    print("")
    #Filter control junctions
      #Take list of all junctions found in controls --> to data.frame with 5 columns (1 for each control sample) for       each
      #junction. Include the number of reads that support that junction for each sample, entering a 0 if it doesn't       appear
      AllGOIControlJunctions_subset<-lapply(AllGOIControlJunctions, function(x){x[,6:7]})
      for (i in 1:length(AllGOIControlJunctions_subset)){
        colnames(AllGOIControlJunctions_subset[[i]])<- c(paste("Score_", i,sep=""), "Name")
      }
      
      AllGOIControlJunctions_Merged<- merge(AllGOIControlJunctions_subset[[1]], 
                                            AllGOIControlJunctions_subset[[2]], by="Name", all=TRUE)
      AllGOIControlJunctions_Merged<- merge(AllGOIControlJunctions_Merged, 
                                            AllGOIControlJunctions_subset[[3]], by="Name", all=TRUE)
      AllGOIControlJunctions_Merged<- merge(AllGOIControlJunctions_Merged, 
                                            AllGOIControlJunctions_subset[[4]], by="Name", all=TRUE)
      AllGOIControlJunctions_Merged<- merge(AllGOIControlJunctions_Merged, 
                                            AllGOIControlJunctions_subset[[5]], by="Name", all=TRUE)
      #Convert NA to 0
      AllGOIControlJunctions_Merged[is.na(AllGOIControlJunctions_Merged)]<- 0
      #Add column with maximum score for each junction
      AllGOIControlJunctions_Merged$Max<- rowMax(as.matrix(AllGOIControlJunctions_Merged[,2:6]))
      #Add column with Number of samples that have that junction
      AllGOIControlJunctions_Merged$SamplesWithJunction<- rowSums(AllGOIControlJunctions_Merged[,2:6]>0)
      #Get list of junctions to NOT look at: Appears in >1 control samples + supported by > 10 reads
      ControlJunctions_Filtered<- AllGOIControlJunctions_Merged[AllGOIControlJunctions_Merged$Max>ControlsListNReads,]
      ControlJunctions_Filtered<- ControlJunctions_Filtered[ControlJunctions_Filtered$SamplesWithJunction>1,]
      
  ############################################################################################################    
  #Get filtered list of junctions for each sample
  AllGOIJunctions<- AllGOIJunctions[SamplesToUse]
  AllFilteredJunctions<- lapply(AllGOIJunctions, FilterJunctions, 
                                ControlJunctions=ControlJunctions_Filtered, DepthFilter=RawDepthFilter)
  print(paste("Number of junctions after control junction filtering + Raw depth filtering (Filter 2/3):", 
              sum(unlist(lapply(AllFilteredJunctions, nrow)))))
  ############################################################################################################
  #Create control splice site filtered list
    #Split splice sites into separate, unique column
    AllSplitControlJunctions<- lapply(AllGOIControlJunctions, SpliceSiteSplit)
    
    #Filter all control junctions by depth
    AllDepthFilteredControlJunctions<- lapply(AllSplitControlJunctions, 
                                              FilterControlJunctions_BasedOnMeanDepth,
                                              Percentage=ControlSiteFilteringPercentage)
    
    #Get all A and B spice sites (summed) for each sample
    AllAControlSpliceSites<-lapply(AllDepthFilteredControlJunctions, GetCanonicalSpliceSiteSum, "A")
    AllBControlSpliceSites<-lapply(AllDepthFilteredControlJunctions, GetCanonicalSpliceSiteSum, "B")
    
    #Rename columns to include sample info
    for (i in 1:length(AllAControlSpliceSites)){
      colnames(AllAControlSpliceSites[[i]])<- c("SpliceSiteA", paste(i, "MeanScore", sep=":"))
    }
    
    for (i in 1:length(AllBControlSpliceSites)){
      colnames(AllBControlSpliceSites[[i]])<- c("SpliceSiteB", paste(i, "MeanScore", sep=":"))
    }
    
    #Merge all A splice sites into one df, all B into one df
    MergedAControlSpliceSites<- AllAControlSpliceSites %>% Reduce(function(dtf1, dtf2) left_join(dtf1, dtf2, by="SpliceSiteA"), .) %>% na.omit
      
    MergedBControlSpliceSites<- AllBControlSpliceSites %>% Reduce(function(dtf1, dtf2) left_join(dtf1, dtf2, by="SpliceSiteB"), .) %>% na.omit
  print(paste("Number of control canonical splice sites (control splice sites after filter 4):", 
        (nrow (MergedAControlSpliceSites) + nrow (MergedBControlSpliceSites))))
  
  
  ############################################################################################################
  #Filter CM junctions based on canonical splice sites
  AllSplitJunctions<- lapply(AllFilteredJunctions, SpliceSiteSplit)

  #Add annotation detailing whether each side of the junction is "canonical" and what the mean depth for that
  #canonical junction is (in controls)
  AllCanonicalAnnotatedJunctions<- lapply(AllSplitJunctions, FilterOutBothNeitherCanonical,
                                          MergedAControlSpliceSites,MergedBControlSpliceSites)
  
  #Select just junctions where exactly one side is canonical
  AllOneCanonicalJunctions<- lapply(AllCanonicalAnnotatedJunctions, function(x) {x[x$OneCanonical=="TRUE",]})
  
  print(paste("Number of CM junctions after filtering based on canonical splice sites (CM junctions after filter 4):",
        sum(unlist(lapply(AllOneCanonicalJunctions, nrow)))))
  
  ############################################################################################################
  #Add column for norm score
  NormScore<- lapply(AllOneCanonicalJunctions, function(x){
  x$MeanBothSites<- rowSums(x[,16:17])
  x$NormScore<- x$Score/x$MeanBothSites
  return(x)
    })
  #Add column for naming based on sample
  Samples<-Samples[SamplesToUse]
  for (i in 1:length(NormScore)){
  NormScore[[i]]$SampleID<- Samples[i]
}

  #merge
  multi_full<- do.call(rbind,NormScore)

  print(paste("Number of CM junctions with Norm score > 0.1 (Filter 5):", nrow(multi_full[multi_full$NormScore>0.1,])))
  print(paste("Number of CM junctions with Norm score > 0.01 (Filter 5):", nrow(multi_full[multi_full$NormScore>0.01,])))
  
  #Test for presence of positive control CM junctions
  print(paste("Number of CM positive control junctions:", nrow((multi_full[multi_full$Name %in% PositiveControlJunctions,]))))
  return(multi_full)
}
```

```{r}
#Testing for differences between old junction analysis (run2) and current (merged)

#Read in data
wd<-"Z:/UKRNASeq/PathogenicSpliceSiteAnalysis/"
run2<- read.csv(paste(wd, "AllFilteredCMJunctions_ControlRemoved5_Depth10_1CanSite_Can.8_Full62GeneList.csv", sep=""))
merged<- read.csv(paste(wd, "MergedAlignment/MERGED_AllFilteredCMJunctions_ControlRemoved5_Depth10_1CanSite_Can.8_Full62GeneList.csv", sep=""))

#Add column with unique ID for junction/sample pair
run2$UniqueID<- paste("Sample", run2$SampleID, "_", run2$Name, sep="")
merged$UniqueID<- paste("Sample", merged$SampleID, "_", merged$Name, sep="")

#Increase in # junctions found
nrow(merged)-nrow(run2) #301

#How many of run2 junctions are found in merged junctions?
nrow(run2[run2$UniqueID %in% merged$UniqueID,]) #76 - only 1 run2 junction was not found in new analysis

#Which run2 junction wasn't found in new analysis? 
notinMerged<-run2[!(run2$UniqueID %in% merged$UniqueID),7] #in sample 19- not of interest
Sample19_Merged<- AllGOIJunctions[[19]]
Sample19_Merged[grep(notinMerged, Sample19_Merged$Name),] #Found in unfiltered list
Sample19_Filtered<- AllFilteredJunctions[[18]]
Sample19_Filtered[grep(notinMerged, Sample19_Filtered$Name),] #Not in canonical filtered
Sample19_Filtered<- AllFilteredJunctions[[19]]
Sample19_Filtered[grep(notinMerged, Sample19_Filtered$Name),] #Found in control filtered
```



#Get filtered list using alternative filtering: based on mean read depth
#Currently not in use
```{r}
AllFilteredJunctions<- lapply(AllGOIJunctions, FilterJunctions_BasedOnMeanDepth, ControlJunctions=ControlJunctions_Filtered,.01)

unlist(lapply(AllFilteredJunctions, nrow))
```

Merge data together for export
```{r}
#In each df, select just score, name, and gene id
JunctionsSubset<- lapply(AllFilteredJunctions,function(x){x[,5:8]})

#Get rid of any samples with no junctions of interest #Only gets rid of sample 10
SamplesSubset<-Samples[lapply(JunctionsSubset, nrow)>0]
JunctionsSubset<- JunctionsSubset[lapply(JunctionsSubset, nrow)>0]

#Add new column with gene ID
for (i in 1:length(JunctionsSubset)){
  JunctionsSubset[[i]]$Sample<- SamplesSubset[i]
}

multi_full<- do.call(rbind,JunctionsSubset)
```


```{r}
write.csv(multi_full, paste(wd, "AllSamples_PotentialPathogenicJunctions_FilteredControlsRemoved_DepthBasedOnMean.01_TTN.LMNA.MYH6.MYH7.ACTN2.csv", sep=""))
```





