---
title: "R Notebook"
output: html_notebook
---

Import libraries:
```{r}
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(metaseqR))
suppressPackageStartupMessages(library(ggVennDiagram))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(tibble))
suppressPackageStartupMessages(library(fgsea))
suppressPackageStartupMessages(library(dplyr))
```




Functions
```{r}
#Create DESeq2 object, check that metadata is correct, and run DESeq2
RunDESeq2<- function(counts, meta){
  #check that counts/meta data has samples in same order
  print(all(rownames(meta) %in% colnames(counts)))
  print(all(rownames(meta) == colnames(counts)))
  
  #create DESeq2 object
  dds<- DESeqDataSetFromMatrix(countData=counts, 
                               colData=meta, 
                               design=~sampleGroup)
  
  #define levels- so that controls will always be used as reference group
  dds$sampleGroup<- relevel(dds$sampleGroup, ref= "control")
  
  #run DE
  dds<- DESeq(dds)
  
  return(dds)

}


#Normalize with variance stabilizing transformation, run and plot PCA
PlotMyPCA<- function(dds, PC1Variance, PC2Variance){
  vsd <- vst(dds, blind=FALSE)
  PCAData<-DESeq2::plotPCA(vsd, intgroup="sampleGroup", returnData=TRUE)
  PCAData$sampleGroup<- gsub("pos", "TTN+", PCAData$sampleGroup)
  PCAData$sampleGroup<- gsub("neg", "TTN-", PCAData$sampleGroup)
  
  p<-ggplot(PCAData, aes(x=PC1, y=PC2, color=sampleGroup))+
    geom_point(size=3)+
    xlab(paste("PC1:", PC1Variance))+
    ylab(paste("PC2:", PC2Variance))+
    ggtitle("PCA: Controls vs TTN- vs TTN+")+
    geom_text(aes(label=colnames(counts(dds)),hjust=0, vjust=0))+
    theme_minimal()
  return(p)
    
  
}

#function that takes dds object and desired contrasts as input, outputs ordered results

ContrastDGE<- function(contrasts, dds){
  ContrastsUse<- c("sampleGroup", contrasts)
  ContrastsResults<- na.omit(results(dds, contrast=ContrastsUse))
  resOrdered<- ContrastsResults[order(ContrastsResults$padj),]
  return(resOrdered)
}

#Get sig DEGs based on 3 filters. Prints out # of DEGs fitting that criteria
GetSigDEGs<- function(DEGs, sigThresh, BasemeanThresh, Log2FCThresh){
        df<-DEGs[DEGs$padj<sigThresh & DEGs$baseMean>BasemeanThresh & abs(DEGs$log2FoldChange) > Log2FCThresh,]
        print(nrow(df))
        return(df)
}

#Run pathway analysis with gsea, plot
RunPlotGSEA<-function(res, gmt, nPermutations, OnlyPlotSig){
        #convert to vector
        ranks <- deframe(res)
        
        # Load the pathways into a named list
        pathways.hallmark <- gmtPathways(paste("C:/Users/Jenna/Documents/UKy_RNASeq/mSigdb/", gmt, sep=""))
        
        #run gsea with x permutations
        fgseaRes <- fgsea(pathways=pathways.hallmark, stats=ranks, nperm=nPermutations)
        
        #tidy up results
        fgseaResTidy <- fgseaRes %>%
        as_tibble() %>%
        arrange(desc(NES))
        
        #Only plot sig pathways, if desired
        if (OnlyPlotSig=="TRUE"){
                fgseaResTidy<-fgseaResTidy[fgseaResTidy$padj<0.05,]
        } else {
                fgseaResTidy<-fgseaResTidy
        }
                
        
        #plot results
        p<-ggplot(fgseaResTidy, aes(reorder(pathway, NES), NES)) +
        geom_col(aes(fill=padj<0.05)) +
        coord_flip() +
        labs(x="Pathway", y="Normalized Enrichment Score",
             title=paste("pathways from GSEA:", gmt, "/", nPermutations, " Permutations")) + 
        theme_minimal()
        
        return(list(p,as.data.frame(fgseaResTidy)))
}
```


#Import raw counts for individual samples (difference sources, described below) and merge into one dataset
```{r}
#Three types of data: 
  #1: samples run in both run 1 and run 2, fastqs merged --> alignment
  #2: samples run in only run 1 (10 and 14)
  #3: samples run in only run 2 (controls)

extension<- ".counts.GENES"

############
#Dataset 1 #
############

#set list of files- merged files
merged_readwd<- "E:/UKy/MergingRNASeq/MergedRuns_RawCounts/"
merged_samples<- c(1:9,11:13,15:19)
merged_files<-c(paste(merged_samples, "_Merged", extension,sep=""))

#Read in counts data
merged_AllCounts<-list()
for (i in 1:length(merged_files)){
  file<- read.table(paste(merged_readwd, merged_files[i], sep=""))
  colnames(file)<- c("GeneID", paste("X", merged_samples[i], sep=""))
  merged_AllCounts[[i]]<- file
}

############
#Dataset 2 #
############

#set list of files- run 1 files (10 and 14)
run1_readwd<- "//Cifs2/rcdata$/UKRNASeq/RawCounts/ORIGINALRUN_NotInUse/"
run1_samples<- c(1:19) #S2 included for now- will change to merged set later
run1_files<-c(paste(run1_samples, extension,sep=""))

#Read in counts data
run1_AllCounts<-list()
for (i in 1:length(run1_files)){
  file<- read.table(paste(run1_readwd, run1_files[i], sep=""))
  colnames(file)<- c("GeneID", paste("X", run1_samples[i], sep=""))
  run1_AllCounts[[i]]<- file
}

############
#Dataset 3 #
############

#set list of files- run 2 files (controls)
run2_readwd<- "//Cifs2/rcdata$/UKRNASeq/RawCounts/"
run2_samples<- c(c(1:9, 11:13, 15:19),paste("UK-Control-", c(1:5), sep=""))
run2_files<-c(paste(run2_samples, extension,sep=""))

#Read in counts data
run2_AllCounts<-list()
for (i in 1:length(run2_files)){
  file<- read.table(paste(run2_readwd, run2_files[i], sep=""))
  colnames(file)<- c("GeneID", paste("X", run2_samples[i], sep=""))
  colnames(file)[2]<- gsub("XUK", "UK",colnames(file)[2]) #removing "X" in column name, if it's for a control
  run2_AllCounts[[i]]<- file
}


#Select just the samples from run 1 and run 2 to be used- that is, NON-MERGED samples. Needs to be at least Samples 10/14 from run 1 and controls from run 2
#Run 1:
run1_PickSamples<- paste("X", c(10,14),sep="")
run1_SelectedSamples<- unlist(lapply(run1_AllCounts, function(x) {colnames(x)[2] %in% run1_PickSamples}))
run1_SelectedSamples_Counts<- run1_AllCounts[run1_SelectedSamples]

#Run 2:
run2_PickSamples<- paste("UK-Control-", c(1:5), sep="")
run2_SelectedSamples<- unlist(lapply(run2_AllCounts, function(x) {colnames(x)[2] %in% run2_PickSamples}))
run2_SelectedSamples_Counts<- run2_AllCounts[run2_SelectedSamples]

#Merge all counts together into one list, then into one DF
AllCounts<- c(merged_AllCounts, run1_SelectedSamples_Counts, run2_SelectedSamples_Counts)
counts<- AllCounts %>%
    Reduce(function(dtf1,dtf2) left_join(dtf1,dtf2,by="GeneID"), .)

#write to csv for later use
write.csv(counts, paste(run2_readwd, "RawCounts_Run1S10.14_Run2Controls_AllOthersMerged.csv", sep=""))

```




Read in counts data (generated above), format, filter out very low counts nd "no_feature", and perform downsampling
```{r}
#Set wds
readwd<-"C:/Users/Jenna/Documents/UKy_RNASeq/RawCounts/"

#read in count data
countsInput<- read.csv(paste(readwd, "RawCounts_Run1S10.14_Run2Controls_AllOthersMerged.csv", sep =""))

countsInput<-countsInput[,c(1:11, 20, 12:14,21,15:19, 22:26)]

#reformat Counts data
rownames(countsInput)<- countsInput$GeneID
countsInput<- countsInput[,3:ncol(countsInput)]
```

```{r}
#remove sample 10 and 14- too low of counts
counts<- countsInput[,!(colnames(countsInput)%in% c("X10", "X14"))]

#Remove highest expressed "gene" - no_feature. Could throw off DE
counts<-counts[!(rownames(counts) %in% "no_feature"),]

#Get rid of any genes with NA values, or where counts are 0 for that gene across all samples
counts<- na.omit(counts) #no NA values
counts<- counts[rowSums(counts)>0,] 

#downsample: set seed so results are replicable
counts<- downsample.counts(counts, seed=42) 
```


Read in metadata
```{r}
#Read in metadata info
wd<- "C:/Users/Jenna/Documents/UKy_RNASeq/"
meta<- read.csv(paste(wd, "MetaData_AllSamples.csv",sep=""))
rownames(meta)<- meta$Sample
meta<- meta[,-1]
meta<- meta[!(rownames(meta) %in% c(paste("X", c(10,14), sep=""))),] 

#Change sample 3/4 to pos
meta[3,2]<- "pos"
meta[4,2]<- "pos"
```

Run DESeq2 on downsampled counts
```{r}
#DESeq2 on downsampled counts- pos vs neg vs controls
dds<- RunDESeq2(counts,meta)
```

Run and plot PCA for raw and DS counts
```{r}
PCA_DS<- PlotMyPCA(dds, "25%", "18%") + theme_classic()
```


```{r}
#Extract comparisons for 3 desired groups
DGERes_controlpos<- ContrastDGE(c("control", "pos"), dds)
DGERes_controlneg<- ContrastDGE(c("control", "neg"), dds)
DGERes_posneg<- ContrastDGE(c("pos", "neg"), dds)

nrow(DGERes_controlpos[DGERes_controlpos$padj<0.05,])
nrow(DGERes_controlneg[DGERes_controlneg$padj<0.05,])
nrow(DGERes_posneg[DGERes_posneg$padj<0.05,])
```

#Plot heatmap
```{r}
#rld<- rlog(dds)
topn<- 150

#select top DE genes for control vs pos or neg and pos vs neg
topgenes<- unique(c(rownames(DGERes_controlneg[1:topn,]), rownames(DGERes_controlpos[1:topn,]), rownames(DGERes_posneg[1:topn,])))
topgenes<-rownames(DGERes_posneg[1:topn,])

#normalize rld counts based on rowMeans
mat<- assay(rld)[topgenes,]
mat<- mat - rowMeans(mat)

#settings for heatmap
subtractFromMin<- -.5
subtractFromMax<- 2
lengthOut<- 100
mat_breaks<- seq(min(mat-subtractFromMin), max(mat-subtractFromMax), length.out=lengthOut)

#Create metadata- to be used to color code to show which group the sample belongs to
metadata<- as.data.frame(meta[,2])
rownames(metadata)<-rownames(meta)
colnames(metadata)<-"sampleGroup"

#plot heatmap
pheatmap(mat, breaks = mat_breaks, color =inferno(length(mat_breaks-1)), show_rownames = FALSE, annotation_col = metadata, border_color=NA)
```


#Performing random sample permutation to show (hopefully) that our groups give the highest number of DEGs
```{r}
AllN_DEGs<-list()
count=1
while (count<=20){
        #Randomly assign each sample to "pos" or "neg"
        CMSamples<-meta[meta$sampleGroup!="control",]
        CMSamples_Random<- CMSamples
        CMSamples_Random$sampleGroup<-sample(CMSamples$sampleGroup)
        
        #merge back with control samples
        meta_random<- rbind(CMSamples_Random, meta[meta$sampleGroup=="control",])
        
        #Perform DE using this new metadata
        dds_random<- RunDESeq2(counts,meta_random)
        
        #Extract comparisons for pos vs neg
        DEGs<-ContrastDGE(c("pos", "neg"), dds_random)
        SigDEGs<- DEGs[DEGs$padj<0.05,]
        AllN_DEGs[count]<-nrow(SigDEGs)
        
        print(count)
        count=count+1
}


```


#Get DEGs specific to + and -, and DEGs common to +/-
```{r}
#Get sig DEGs- 3 potential thresholds
sigThres<-0.05
BasemeanThresh<-0
Log2FCThresh<-0.0

tPosDEGs<- GetSigDEGs(DGERes_controlpos, sigThres, BasemeanThresh, Log2FCThresh) 
tNegDEGs<- GetSigDEGs(DGERes_controlneg, sigThres, BasemeanThresh, Log2FCThresh)

#Get DEGs specific to tPos or tNeg, or common to tPos/tNeg
tPosSpecific<-tPosDEGs[!(rownames(tPosDEGs) %in% rownames(tNegDEGs)),] #573 (p<0.05) / 307 (p<0.01)
nrow(tPosSpecific)
tNegSpecific<-tNegDEGs[!(rownames(tNegDEGs) %in% rownames(tPosDEGs)),] #2,361 (p<0.05) / 1,668 (p<0.01)
nrow(tNegSpecific)
CommonDEGs<-tPosDEGs[rownames(tPosDEGs) %in% rownames(tNegDEGs),]
nrow(CommonDEGs) #2,343 p<0.05 / 1,165 p<0.01

venn<-list(tPos=rownames(tPosDEGs), tNeg=rownames(tNegDEGs))
ggVennDiagram(venn)
```

For p<0.05 thresh: #samples in tPos/tNeg is split ~60/40. But the # of unique DEGs is split ~80/20. That is- we see a lot fewer tPos specific DEGs than we might expect, if the groups were totally comparable. Maybe suggests that the tPos are more similar to one another than the tNeg. Also can see this in the shear number of DEGs found in tPos/tNeg.

#Plotting heatmap of tPos and tNeg specific DEGs only
```{r}
#Select just top DEGs
Topn<-2000
topgenes<- unique(c(rownames(tPosSpecific[1:topn,]), rownames(tNegSpecific[1:topn,])))
#topgenes<-unique(rownames(tPosSpecific), rownames(tNegSpecific)) #Option to plot all

#normalize rld counts based on rowMeans
mat<- assay(rld)[topgenes,]
mat<- mat - rowMeans(mat)

#settings for heatmap
subtractFromMin<- -.7
subtractFromMax<- 1.8
lengthOut<- 100
mat_breaks<- seq(min(mat-subtractFromMin), max(mat-subtractFromMax), length.out=lengthOut)

#Create metadata- to be used to color code to show which group the sample belongs to
metadata<- as.data.frame(meta[,2])
rownames(metadata)<-rownames(meta)
colnames(metadata)<-"sampleGroup"

#plot heatmap
pheatmap(mat, breaks = mat_breaks, color =inferno(length(mat_breaks-1)), show_rownames = FALSE, annotation_col = metadata, border_color=NA)
```


#For DEgs found in both tPos and tNeg- look at pattern of expression. Expect it to be very similar.
```{r}
#Separate counts for tPos and tNeg samples
tPosCounts<- as.data.frame(rowMeans(counts[,colnames(counts) %in% as.character(meta[meta$sampleGroup=="pos",1])]))
tNegCounts<- as.data.frame(rowMeans(counts[,colnames(counts) %in% as.character(meta[meta$sampleGroup=="neg",1])]))

#Merge counts into 1 df- for each DEG in both, mean raw counts for tPos and tNeg
CommonDEGs_counts<-merge(as.data.frame(CommonDEGs), tPosCounts, by=0)[,c(1,8)]
rownames(CommonDEGs_counts)<-CommonDEGs_counts$Row.names
CommonDEGs_counts<- merge(CommonDEGs_counts, tNegCounts, by=0)[,2:4]
colnames(CommonDEGs_counts)<-c("CommonDEGs", "Mean_tPos", "Mean_tNeg")
rownames(CommonDEGs_counts)<-CommonDEGs_counts$CommonDEGs
CommonDEGs_counts<- CommonDEGs_counts[,-1]

#Optional: transformation
CommonDEGs_log2<- log2(CommonDEGs_counts+0.0001)
CommonDEGs_CPM<- (CommonDEGs_counts/colSums(counts)) * 1000000
CommonDEGs_CPM_log2<- log2(CommonDEGs_CPM+0.0001)

df<-CommonDEGs_CPM_log2

p<-ggplot(df, aes(x=Mean_tPos, y=Mean_tNeg))+
        geom_point()
```



#Pathway analysis using fgsea
```{r}
#database options I've downloaded
PathwayOptions<-c("h.all.v7.1.symbols.gmt",
                  "c2.cp.kegg.v7.1.symbols.gmt",
                  "c3.all.v7.1.symbols.gmt",
                  "c5.all.v7.1.symbols.gmt")

#Reformat list of DEGs in tPos only: only need gene ID and stat
tPosPath<- as.data.frame(tPosSpecific[,c(2,4,6)])
tPosPath$geneID<-rownames(tPosPath)
tPosPathInput<-tPosPath[,c(4,2)]

#Reformat list of DEGs in tNeg only: only need gene ID and stat
tNegPath<- as.data.frame(tNegSpecific[,c(2,4,6)])
tNegPath$geneID<-rownames(tNegPath)
tNegPathInput<-tNegPath[,c(4,2)]

#Reformat list of DEGs in both tPos and tNeg: only need gene ID and stat
CommonDEGsPath<- as.data.frame(CommonDEGs[,c(2,4,6)])
CommonDEGsPath$geneID<-rownames(CommonDEGsPath)
CommonDEGsPathInput<-CommonDEGsPath[,c(4,2)]
        

#Running pathway analysis
gsea<-RunPlotGSEA(CommonDEGsPathInput, PathwayOptions[1], 10000, "TRUE")

#Hallmark database: saving for later
tNeg_Hallmark<-RunPlotGSEA(tNegPathInput, PathwayOptions[1], 10000, "TRUE")[[2]]
tPos_Hallmark<-RunPlotGSEA(tPosPathInput, PathwayOptions[1], 100000, "TRUE")[[2]]

#Running for all 4 databases. To combine and export tables.
CommonDEGs_SigPathways<-lapply(PathwayOptions, function(x){RunPlotGSEA(CommonDEGsPathInput, x, 10000, "TRUE")[[2]]})
all<-do.call("rbind", CommonDEGs_SigPathways) #110 pathways
all$leadingEdge<-as.character(all$leadingEdge)

tNeg_SigPathways<-lapply(PathwayOptions, function(x){RunPlotGSEA(tNegPathInput, x, 10000, "TRUE")[[2]]})
all<-do.call("rbind", tNeg_SigPathways) #517 pathways
all$leadingEdge<-as.character(all$leadingEdge)

write.csv(all, paste(wd, "DEGsCommonTo_BothtPostNeg_GSEAResults_FourPathwayDatabases_051320.csv",sep=""))

#Plot Hallmark pathways for tNeg and tPos on one plot
tPos_H<-tPos_Hallmark[,c(1,5,6)]
tNeg_H<-tNeg_Hallmark[,c(1,5,6)]

tPos_H$Group<-"tPos"
tNeg_H$Group<-"tNeg"
All_H<-rbind(tPos_H, tNeg_H)

#plot results
p<-ggplot(All_H, aes(reorder(pathway, NES), NES)) +
        geom_col(aes(fill=Group)) +
        coord_flip() +
        labs(x="Pathway", y="Normalized Enrichment Score") + 
        theme_minimal()
```



