---
title: "miRNA Analysis: DESeq2, Plotting"
output: html_notebook
---
#DE analysis of miRNA counts 

4 DE analyses performed:

1. Controls vs all CM
2. Controls vs TTN+
3. Controls vs TTN-
4. TTN+ vs TTN-


Only the first analysis shows clustering of samples (both in DE heatmap and PCA)

Libraries used
```{r}
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(multiMiR))
suppressPackageStartupMessages(library(metaseqR))
suppressPackageStartupMessages(library(plyr))
```

Functions:
```{r}
ContrastDGE<- function(contrasts, dds){
  ContrastsUse<- c("sampleGroup", contrasts)
  ContrastsResults<- na.omit(results(dds, contrast=ContrastsUse))
  resOrdered<- ContrastsResults[order(ContrastsResults$padj),]
  print(nrow(resOrdered[resOrdered$padj<0.05,]))
  return(resOrdered)
}

RunDESeq2<- function(counts, meta){
  #check that counts/meta data has samples in same order
  print(all(rownames(meta) %in% colnames(counts)))
  print(all(rownames(meta) == colnames(counts)))
  
  #create DESeq2 object
  dds<- DESeqDataSetFromMatrix(countData=counts, 
                               colData=meta, 
                               design=~Sex+sampleGroup)
  
  #define levels- so that controls will always be used as reference group
  dds$sampleGroup<- relevel(dds$sampleGroup, ref= "control")
  
  #run DE
  dds<- DESeq(dds)
  
  return(dds)

}
```

#DESeq2 

Two potential data sources:

1. Basespace miRDeep2 --> mature and isomiR files --> collapse

2. Command line miRDeep2, defualt settings

Read in data
```{r}
#Data source 1- basespace files
        #Set working directory
        wd<- "C:/Users/Jenna/Documents/UKy_RNASeq/RawCounts/"
        
        matureInput<- read.csv(paste(wd, "miRNA_matureHits_downloadedSeparately.csv", sep=""))
        isomirInput<-read.csv(paste(wd, "miRNA_isomirs_downloadedSeparately.csv",sep=""))
        
        #Remove sequence info from isomirs, then collapse isomirs together
        isomirInput$isomirs<- gsub(".*_", "", isomirInput$X)
        isomirInput<-isomirInput[,-1]
        isomir_collapsed<-isomirInput %>% group_by(isomirs) %>%  summarise_each(funs(sum)) %>% as.data.frame()
        rownames(isomir_collapsed)<-isomir_collapsed$isomirs
        isomir_collapsed<- isomir_collapsed[,-1]
        
        #rbind mature and isomirs together. Make sure columns are in the same order
        rownames(matureInput)<-matureInput$X
        matureInput<-matureInput[,-1]
        counts_merged<-rbind(matureInput,isomir_collapsed)
        
        #remove everything after 3rd "-" (-1, -5p, etc), then merge and sum together based on new miR IDs
        counts_collapsedInput<-counts_merged
        counts_collapsedInput$geneIDs<- sub("^([^-]*-[^-]*-[^-]*).*", "\\1", rownames(counts_collapsedInput))
        counts_collapsedInput$geneIDs<- gsub("mir", "miR", counts_collapsedInput$geneIDs)
                
        #Collapse isomirs and mature toegether
        counts_collapsed<-counts_collapsedInput %>% group_by(geneIDs) %>% summarise_each(funs(sum)) %>% as.data.frame
        rownames(counts_collapsed)<-counts_collapsed$geneIDs
        counts_collapsed<- counts_collapsed[,2:ncol(counts_collapsed)]
        rownames(counts_collapsed)<- rownames(counts_collapsed)
        
#Data source 2- command line miRDeep2. Collapsed into 1 file in "ReofrmatingRawmiRDeep2Output.Rmd"
                wd<-"E:/UKy/miRNA_Bams/SmallRNA_092019/"
                CLcountsInput<-read.csv(paste(wd, "AllCountsFromCL_Merged.csv",sep=""), row.names = "geneID" )
                CLcountsInput<- CLcountsInput[,-1]
                CLcounts<-as.data.frame(apply(CLcountsInput, 2, as.integer))
                rownames(CLcounts)<-rownames(CLcountsInput)
                
#Select which data source to use for the rest of the analysis. CL counts used downstream.
counts<-CLcounts
```

#Comparing TTN+ vs TTN- vs Controls
```{r}
#process, rename samples
hits<- counts
colnames(hits)<- gsub("counts_Control", "Control", gsub("counts_Sample", "X", colnames(hits)))

#get rid of any IDs that have 0 counts for all samples 
hits<- hits[rowSums(hits)>0,]


#set sample groups
sampleGroup<- c("pos", "pos", "pos", "pos", "pos", "neg", "neg", "neg", "pos", "neg", "neg", "neg", "neg", "neg", "neg", "neg", "neg", rep("control", 5)) 
```


```{r}
#create metadata
sample<- colnames(hits)
meta<- data.frame(sample, sampleGroup)
rownames(meta)<- meta$sample
```

```{r}
detach("package:metaseqR")
detach("package:DESeq")
```

#Plot on PCA- need to create DESeq2 object first
```{r}
#Create DESeq2 object
dds<- DESeqDataSetFromMatrix(countData=hits, 
                             colData=meta, 
                             design=~sampleGroup)
dds$sampleGroup<- relevel(dds$sampleGroup, ref= "control")

#Normalize using variance stabilizing transformation- DESeq2 function
vsd <- varianceStabilizingTransformation(dds, blind=FALSE)
PCAData<-plotPCA(vsd, intgroup="sampleGroup", returnData=TRUE)

#Rename samples from pos --> TTN+ and neg --> TTN-
PCAData$sampleGroup<- gsub("pos", "TTN+", PCAData$sampleGroup)
PCAData$sampleGroup<- gsub("neg", "TTN-", PCAData$sampleGroup)

#plot PCA
p<-ggplot(PCAData, aes(x=PC1, y=PC2, color=sampleGroup))+
  geom_point(size=3)+
  xlab("PC1: 21% Variance")+
  ylab("PC2: 13% Variance")+
  ggtitle("PCA: Controls vs TTN- vs TTN+")+
  geom_text(aes(label=rownames(PCAData)),hjust=.5, size= 5,vjust=-1.2, color="black")+
  theme_classic()

p

```

Samples slightly cluster by CM status, but do NOT cluster based on TTN mutations

Run differential expression on this DESeq2 object- tpos vs tneg vs controls
```{r}
dds<- DESeq(dds)

res_tPosvControl<- ContrastDGE(c("control", "pos"), dds)
res_tNegvControl<-ContrastDGE(c("control", "neg"), dds)
res_tPosvtNeg<-ContrastDGE(c("pos", "neg"), dds)


#Normalize
rld<- rlog(dds)

#select top genes: all sig DE miRNAs
topgenes<- unique(c(rownames(res_tPosvControl[res_tPosvControl$padj<0.05,]), 
             rownames(res_tNegvControl[res_tNegvControl$padj<0.05,]),
             rownames(res_tPosvtNeg[res_tPosvtNeg$padj<0.05,])))

#Alt: select top in each, even if not necessarily DE
topn<-10
topgenes<- unique(c(rownames(res_tPosvControl[1:topn,]), 
             rownames(res_tNegvControl[1:topn,]),
             rownames(res_tPosvtNeg[1:topn,])))

#settings for heatmap
subtractFromMin<- -.7
subtractFromMax<- 2.75
lengthOut<- 100

mat<- assay(rld)[topgenes,]
mat<- mat - rowMeans(mat)
mat_breaks<- seq(min(mat-subtractFromMin), max(mat-subtractFromMax), length.out=lengthOut)


#Create metadata- to be used to color code to show which group the sample belongs to
metadata<- data.frame(Sample_Group=PCAData$sampleGroup, row.names=PCAData$name)

#plot heatmap
pheatmap(mat, breaks = mat_breaks, color =inferno(length(mat_breaks-1)), show_rownames = FALSE, annotation_col = metadata)
```

#Rerun the above analysis, comparing ALL CM vs controls

Create DESeq2 object and PCA plot
```{r}
#set sample groups
sampleGroup<- c(rep("CM", 17), rep("control", 5)) 

#create metadata
sample<- colnames(hits)
meta<- data.frame(sample, sampleGroup)
rownames(meta)<- meta$sample

#create DESeq2 object
dds<- DESeqDataSetFromMatrix(countData=hits, 
                             colData=meta, 
                             design=~sampleGroup)

dds$sampleGroup<- relevel(dds$sampleGroup, ref= "control")

#Normalize using variance stabilizing transformation- DESeq2 function
vsd <- varianceStabilizingTransformation(dds, blind=FALSE)
PCAData<-plotPCA(vsd, intgroup="sampleGroup", returnData=TRUE)

#Rename samples from pos --> TTN+ and neg --> TTN-
PCAData$sampleGroup<- gsub("pos", "TTN+", PCAData$sampleGroup)
PCAData$sampleGroup<- gsub("neg", "TTN-", PCAData$sampleGroup)

#Plot PCA
p<-ggplot(PCAData, aes(x=PC1, y=PC2, color=sampleGroup))+
  geom_point(size=3)+
  xlab("PC1: 21% Variance")+
  ylab("PC2: 13% Variance")+
  ggtitle("PCA: Controls vs TTN- vs TTN+")+
  theme_classic()

p
```

Run differential expression on this DESeq2 object- CM vs Controls
```{r}
#Run DESeq2
dds<- DESeq(dds)

#extract results
res<- as.data.frame(results(dds))
res<- na.omit(res)
res<- res[order(res$padj),]

nrow(res[res$padj<0.05,])#30 DE genes

sig<-res[res$padj<0.05 & abs(res$log2FoldChange>0.5),] #18
miRNA_sig<-res[res$padj<0.05,] #39
summary(results(dds))
```

Plot Heatmap for these DE results

```{r}
#Normalize
rld<- rlog(dds)

#select top X DE genes for control vs CM
#topgenes<- rownames(res[1:,])
topgenes<-rownames(sig)

#settings for heatmap
subtractFromMin<- -1.1
subtractFromMax<- .8
lengthOut<- 100

mat<- assay(rld)[topgenes,]
mat<- mat - rowMeans(mat)
mat_breaks<- seq(min(mat-subtractFromMin), max(mat-subtractFromMax), length.out=lengthOut)


#Create metadata- to be used to color code to show which group the sample belongs to
metadata<- data.frame(Sample_Group=PCAData$sampleGroup, row.names=PCAData$name)

#plot heatmap
pheatmap(mat, breaks = mat_breaks, color =inferno(length(mat_breaks-1)), show_rownames = FALSE, annotation_col = metadata)
```

Write DE results to csv
```{r}
write.csv(as.data.frame(res), paste(wd, "DESeq2/miRNADESeq2Results_matureInput.csv", sep =""))
```


Volcano plot on this data
```{r}
#convert DE results to dataframe, select only DE miRNAs (here set as < 0.1 to visualize more)
CMvCon<- as.data.frame(res)
threshold_OE<- CMvCon$padj<0.1
CMvCon$threshold<- threshold_OE

#Plot
p<-ggplot(CMvCon) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj), colour=threshold)) +
        ggtitle("Controls vs CM: miRNA DE") +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        #scale_y_continuous(limits = c(0,50)) +
        theme(legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25)))+
        scale_color_manual(values= c("black", "red"))+
        theme_classic()+
        theme(legend.position="none")
p
```


#Performing random sample permutation to show (hopefully) that our groups give the highest number of DEGs
```{r}
AllN_DEGs<-list()
All_SampleGroups<-list()
count=1
while (count<=10){
        #Randomly assign each sample to "pos" or "neg"
        meta_random<- meta
        meta_random$sampleGroup<-sample(meta_random$sampleGroup)
     
        #Perform DE using this new metadata
        dds_random<- RunDESeq2(hits,meta_random)
        
        #Extract comparisons for pos vs neg
        DEGs<-ContrastDGE(c("CM", "control"), dds_random)
        SigDEGs<- DEGs[DEGs$padj<0.05,]
        AllN_DEGs[count]<-nrow(SigDEGs)
        All_SampleGroups[[count]]<-meta_random
        
        print(count)
        count=count+1
}
```



#################################
using isomirs, other combinations
#################################

Everything below not currently in use, but I've retained it here to see previous analyses as needed


Read in data
```{r}
wd<- "//Cifs2/rcdata$/UKRNASeq/miRNA/"
hitsInput<- read.delim(paste(wd, "RawData/Hits.txt", sep=""))
hitsInput<- read.delim(paste(wd, "RawData/FilteredIsoHits.txt", sep=""))
```

#Create PCA plots

```{r}
#Function that takes hit counts and metadata as input, outputs PCA plot
myPCA<- function(hits, meta){
  if (all(rownames(meta) %in% colnames(hits))){
    if (all(rownames(meta) == colnames(hits))){
      dds<- DESeqDataSetFromMatrix(countData=hits, 
                             colData=meta, 
                             design=~sampleGroup)
      dds$sampleGroup<- relevel(dds$sampleGroup, ref= "control")
      rld<- rlog(dds)
      plotPCA(rld, intgroup="sampleGroup")
    }
  } else {
    print ("Error: check metadata")
  }
}
```


Standard: all samples (that are also in RNASeq)/3 groups
```{r}
dds<- DESeqDataSetFromMatrix(countData=hits, 
                             colData=meta, 
                             design=~sampleGroup)
dds$sampleGroup<- relevel(dds$sampleGroup, ref= "control")
rld<- rlog(dds)
plotPCA(rld, intgroup="sampleGroup")

myPCA(hits,meta)
```

Try DE using just controls vs CM
```{r}
#set sample groups
sampleGroup<- c(rep("CM", 16), rep("control", 5))

#create metadata
sample<- colnames(hits)
meta<- data.frame(sample, sampleGroup)
rownames(meta)<- meta$sample

myPCA(hits,meta)
```

```{r}
#Throwing out sample 17- very low counts
hits<- hitsFinal[,c(1:13, 15:21)]

#set sample groups
sampleGroup<- c(rep("CM", 15), rep("control", 5))

#create metadata
sample<- colnames(hits)
meta<- data.frame(sample, sampleGroup)
rownames(meta)<- meta$sample

myPCA(hits,meta)
```

Throwing out sample 17 + all 3 groups
```{r}
#Throwing out sample 17- very low counts
hits<- hitsFinal[,c(1:13, 15:21)]

#set sample groups
sampleGroup<- c("pos", "pos", "neg", "pos", "neg", "neg", "neg", "pos", "neg", "neg", "neg", "neg", "neg", "neg", "neg", rep("control", 5))

#create metadata
sample<- colnames(hits)
meta<- data.frame(sample, sampleGroup)
rownames(meta)<- meta$sample

myPCA(hits,meta)
```

Including all samples (except sample 17), not just samples with RNASeq data
```{r}
#Get hit counts for all samples
hits<- hitsInput[,c(1,3:26)]
hits<- hits[-1,]
rownames(hits)<- hits$SampleNumber
hits<- hits[,-1]

colnames(hits)<- c("Control4", "Control3", "Control2","Control1","Control5", "X7", "X14", "X12", "X13", "X11", "X10", "X8", "X5", "X6", "X4", "X16", "X15", "X3", "X19", "X1", "X2", "X9", "X18", "X17")

#get rid of any IDs that have 0 counts for all samples
hits<- hits[rowSums(hits)>0,]

#reorder
hits<- hits[,c(6:23, 1:5)]

#set sample groups
sampleGroup<- c(rep("CM", 18), rep("control", 5))

#create metadata
sample<- colnames(hits)
meta<- data.frame(sample, sampleGroup)
rownames(meta)<- meta$sample

myPCA(hits,meta)
dds<- DESeqDataSetFromMatrix(countData=hits, 
                             colData=meta, 
                             design=~sampleGroup)
dds$sampleGroup<- relevel(dds$sampleGroup, ref= "control")
rld<- rlog(dds)
plotPCA(rld, intgroup="sampleGroup")
```

Repeating for all3 groups
```{r}
#reorder
hits<- hits[,c(6:23, 1:5)]

#set sample groups
sampleGroup<- c(c("neg", "neg", "neg", "neg", "neg", "neg", "neg", "pos", "neg", "neg", "neg", "neg", "neg", "neg", "pos", "pos", "pos", "neg"), rep("control", 5))

#create metadata
sample<- colnames(hits)
meta<- data.frame(sample, sampleGroup)
rownames(meta)<- meta$sample

#myPCA(hits,meta)

dds<- DESeqDataSetFromMatrix(countData=hits, 
                             colData=meta, 
                             design=~sampleGroup)
dds$sampleGroup<- relevel(dds$sampleGroup, ref= "control")
rld<- rlog(dds)
plotPCA(rld, intgroup="sampleGroup")
```

#Differential expression using data that looks tightest in PCA
```{r}
#Using 2 groups, excluding sample 17 and samples without RNASeq data

#Throwing out sample 17- very low counts
hits<- hitsFinal[,c(1:13, 15:21)]

#set sample groups
sampleGroup<- c(rep("CM", 15), rep("control", 5))

#create metadata
sample<- colnames(hits)
meta<- data.frame(sample, sampleGroup)
rownames(meta)<- meta$sample

#Run DESeq2, extract results
dds<- DESeqDataSetFromMatrix(countData=hits, 
                             colData=meta, 
                             design=~sampleGroup)

dds$sampleGroup<- relevel(dds$sampleGroup, ref= "control")

dds<- DESeq(dds)

res<- as.data.frame(results(dds))
res<- na.omit(res)
res<- res[order(res$padj),]

nrow(res[res$padj<0.05,])
#40 DE genes

summary(results(dds))
```

```{r}
head(res, 40)
```

Heatmap
```{r}
rld<- rlog(dds)
topgenes<- rownames(res[1:100,])
mat<- assay(rld)[topgenes,]
mat<- mat - rowMeans(mat)

subtractFromMin<-0
subtractFromMax<- 2.5
lengthOut<- 100
mat_breaks<- seq(min(mat-subtractFromMin), max(mat-subtractFromMax), length.out=lengthOut)
pheatmap(mat, breaks = mat_breaks, color =inferno(length(mat_breaks-1)))
```

DESeq2 with 3 groups
```{r}
#Using 2 groups, excluding sample 17 and samples without RNASeq data

#Throwing out sample 17- very low counts
hits<- hitsFinal[,c(1:13, 15:21)]

#set sample groups
sampleGroup<- c(c("pos","pos","neg", "pos", "neg", "neg","neg", "pos", "neg","neg", "neg", "neg", "neg", "neg", "neg"), rep("control", 5))

#create metadata
sample<- colnames(hits)
meta<- data.frame(sample, sampleGroup)
rownames(meta)<- meta$sample

#Run DESeq2, extract results
dds<- DESeqDataSetFromMatrix(countData=hits, 
                             colData=meta, 
                             design=~sampleGroup)

dds$sampleGroup<- relevel(dds$sampleGroup, ref= "control")

dds<- DESeq(dds)

resPosControl<- results(dds, contrast=c("sampleGroup", "pos", "control"))
resPosControl<- na.omit(resPosControl[order(resPosControl$padj),])
#0 DE

resNegControl<- results(dds, contrast=c("sampleGroup", "neg", "control"))
resNegControl<- na.omit(resNegControl[order(resNegControl$padj),])
nrow(resNegControl[resNegControl$padj<0.05,])
#24 DE

resPosNeg<- results(dds, contrast=c("sampleGroup", "pos", "neg"))
resPosNeg<- na.omit(resPosNeg[order(resPosNeg$padj),])
nrow(resPosNeg[resPosNeg$padj<0.05,])
#0 DE

#heatmap
rld<- rlog(dds)
topgenes<- unique(c(rownames(resPosControl)[1:40], rownames(resNegControl)[1:40], rownames(resPosNeg)[1:40]))
mat<- assay(rld)[topgenes,]
mat<- mat - rowMeans(mat)

subtractFromMin<-0
subtractFromMax<- 2.5
lengthOut<- 100
mat_breaks<- seq(min(mat-subtractFromMin), max(mat-subtractFromMax), length.out=lengthOut)
pheatmap(mat, breaks = mat_breaks, color =inferno(length(mat_breaks-1)))
```

#Running DESeq2 again- collapsing miRNA species into 1

In hits file, there are multiple entries for most species- different isomirs

In previous analysis, these were all considered separately

Here, I'm collapsing all isomirs for each species into one, adding the counts together

Maybe this is stupid. I dont know

```{r}
#process, rename samples correctly
hits<- hitsInput[,c(1,3:26)]
hits<- hits[-1,]
rownames(hits)<- hits$SampleNumber
hits<- hits[,-1]

colnames(hits)<- c("Control4", "Control3", "Control2","Control1","Control5", "X7", "X14", "X12", "X13", "X11", "X10", "X8", "X5", "X6", "X4", "X16", "X15", "X3", "X19", "X1", "X2", "X9", "X18", "X17")

#get rid of any IDs that have 0 counts for all samples
hits<- hits[rowSums(hits)>0,]
#gets rid of 0 hits (?)

#Reorder columns, get rid of any that aren't used in RNASeq
hits<- hits[,c(20,21,18,13,14,6,12,22,10,8,9,17,16,24,23,19,4,3,2,1,5)]

hitsFinal<- hits

hitsFinal$miRNA<- gsub(".*_", "",rownames(hitsFinal))

df<- as_tibble(hitsFinal)
df<-df%>% group_by(miRNA) %>% summarise_at(colnames(hits)[1:21],sum, na.rm=TRUE)
hits<- as.data.frame(df)
rownames(hits)<-hits$miRNA
hits<- hits[,-1]

#Running DESeq2 on this
#Throwing out sample 17- very low counts
hits<- hits[,c(1:13, 15:21)]

#set sample groups
sampleGroup<- c(rep("CM", 15), rep("control", 5))

#create metadata
sample<- colnames(hits)
meta<- data.frame(sample, sampleGroup)
rownames(meta)<- meta$sample

#Run DESeq2, extract results
dds<- DESeqDataSetFromMatrix(countData=hits, 
                             colData=meta, 
                             design=~sampleGroup)

dds$sampleGroup<- relevel(dds$sampleGroup, ref= "control")

dds<- DESeq(dds)

res<- as.data.frame(results(dds))
res<- na.omit(res)
res<- res[order(res$padj),]

nrow(res[res$padj<0.05,])
#18 DE genes

summary(results(dds))
  
```

```{r}
#heatmap
rld<- rlog(dds)
topgenes<- rownames(res[1:75,])
mat<- assay(rld)[topgenes,]
mat<- mat - rowMeans(mat)

subtractFromMin<- 0
subtractFromMax<- 2
lengthOut<- 100
mat_breaks<- seq(min(mat-subtractFromMin), max(mat-subtractFromMax), length.out=lengthOut)
pheatmap(mat, breaks = mat_breaks, color =inferno(length(mat_breaks-1)))
```


#DESeq2: Running on mRNA data. Comparing between just control/CM
Will need this for multiMiR analysis
```{r}

#set sample groups
sampleGroup<- c(rep("CM", 16), rep("control", 5))

#read in count data, format, get rid of NA data, get rid of any genes where all counts are 0
countsInput<- read.csv(paste(wd, "RawCounts/CountData.csv", sep=""))
rownames(countsInput)<- countsInput$X
countsInput<- countsInput[,-1]
colnames(countsInput)<- gsub(".counts.*", "", colnames(countsInput))
counts<- na.omit(countsInput)
counts<- counts[rowSums(counts)>0,]

#create metadata
sample<- colnames(countsInput)
meta<- data.frame(sample, sampleGroup)
rownames(meta)<- meta$sample

#check that counts/meta data has samples in same order
all(rownames(meta) %in% colnames(counts))
all(rownames(meta) == colnames(counts))

#create DESeq2 object
dds<- DESeqDataSetFromMatrix(countData=counts, 
                             colData=meta, 
                             design=~sampleGroup)

#define levels- so that controls will always be used as reference group
dds$sampleGroup<- relevel(dds$sampleGroup, ref= "control")

#run DE
dds<- DESeq(dds)

res<- as.data.frame(results(dds))
res<- res[order(res$padj),]
#9,000 DE

vsd <- vst(dds, blind=FALSE)
plotPCA(vsd, intgroup="sampleGroup")

```

```{r}
#Heatmap
rld<- rlog(dds)
topgenes<- rownames(res[1:200,])
mat<- assay(rld)[topgenes,]
mat<- mat - rowMeans(mat)

subtractFromMin<-0
subtractFromMax<- 2.5
lengthOut<- 100
mat_breaks<- seq(min(mat-subtractFromMin), max(mat-subtractFromMax), length.out=lengthOut)
pheatmap(mat, breaks = mat_breaks, color =inferno(length(mat_breaks-1)))
```

```{r}
#write DEGs to csv
write.csv(res, paste(wd, "DESeq2/Final/DGERes_ControlvsCm.csv", sep =""))
```
