---
title: "miRNA Analysis: DESeq2, Plotting"
output: html_notebook
---
#DE analysis of miRNA counts 

4 DE analyses performed:

1. Controls vs all CM
2. Controls vs TTN+
3. Controls vs TTN-
4. TTN+ vs TTN-


Only the first analysis shows clustering of samples (both in DE heatmap and PCA)

**Libraries used**
```{r}
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(multiMiR))
suppressPackageStartupMessages(library(metaseqR))
suppressPackageStartupMessages(library(plyr))
```

**Functions:**
```{r}
ContrastDGE<- function(contrasts, dds){
  ContrastsUse<- c("sampleGroup", contrasts)
  ContrastsResults<- na.omit(results(dds, contrast=ContrastsUse))
  resOrdered<- ContrastsResults[order(ContrastsResults$padj),]
  print(nrow(resOrdered[resOrdered$padj<0.05,]))
  return(resOrdered)
}

RunDESeq2<- function(counts, meta){
  #check that counts/meta data has samples in same order
  print(all(rownames(meta) %in% colnames(counts)))
  print(all(rownames(meta) == colnames(counts)))
  
  #create DESeq2 object
  dds<- DESeqDataSetFromMatrix(countData=counts, 
                               colData=meta, 
                               design=~Sex+sampleGroup)
  
  #define levels- so that controls will always be used as reference group
  dds$sampleGroup<- relevel(dds$sampleGroup, ref= "Control")
  
  #run DE
  dds<- DESeq(dds)
  
  return(dds)

}

#Plot PCA nicely
MyPCA<- function(dds, metadata, PC1Variance, PC2Variance, Groupcolors,miny, maxy, minx, maxx, pointSize, intgroup) {
  #Normalized (variance stabilized transformation), then run PCA
  vsd <- varianceStabilizingTransformation(dds, blind=TRUE)
  PCA<-plotPCA(vsd, intgroup=intgroup,returnData=TRUE)
  
  #plot
  p<- ggplot(PCA, aes(x=PC1, y=PC2, color=group))+
  geom_point(size=pointSize)+
  scale_color_manual(values=Groupcolors)+
  theme_classic()+
  ylim(miny,maxy)+
  xlim(minx,maxx)+
  ylab(paste("PC2: ", PC2Variance, "% Variance", sep =""))+
  xlab(paste("PC1: ", PC1Variance, "% Variance", sep =""))+
  theme(axis.text.y   = element_text(size=15, colour = "black", face="bold"),
        axis.title.y  = element_text(size=18, colour = "black", face="bold"),
        axis.title.x  = element_text(size=18, colour = "black", face="bold"),
        axis.text.x  = element_text(size=15, colour = "black", face="bold"),
        legend.background = element_rect(fill=NA, size=1, 
                                         linetype="solid", color="black"),
        legend.title=element_blank(),
        legend.text = element_text(size=18, face="bold"),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"),
        legend.position = c(.85,.9),
        panel.border = element_rect(colour = "black", fill=NA, size=2)
        )
  return(p)
}

#Plot heatmap: takes as input DE results (padj ordered lowest to highest, rlog transformed data, metadata (blocker status and Sample type), top genes to include, and min and max subtract (used for heatmap aesthetics))
PlotHeatmap<- function(rld, metadata, topgenes, 
                       MinSubtract, MaxSubtract, fontsize, annotationColors){
  mat<- assay(rld)[topgenes,]
  mat<- mat - rowMeans(mat)
  colnames(mat)<-metadata$Sample
  
  #Heatmap settings
  mat_breaks<- seq(min(mat-MinSubtract), max(mat-MaxSubtract), length.out=75)

  #Create metadata- to be used to color code to show which group the sample belongs to
  metadata<- data.frame(Group=metadata$sampleGroup, 
                        row.names=metadata$Sample)

  #plot heatmap
  p<-pheatmap(mat, breaks = mat_breaks, 
            color =colorRampPalette( c("red", "black", "green"), space="rgb")(100),
            show_rownames = TRUE, show_colnames=TRUE,
            annotation = metadata,
            annotation_colors= annotationColors,
            fontsize_row = fontsize,fontsize=12,treeheight_row=0,
            border_color= NA,
            clustering_distance_cols = "euclidean")
  return(p)
}

```

#DESeq2 

Two potential data sources:

1. Basespace miRDeep2 --> mature and isomiR files --> collapse

2. Command line miRDeep2, defualt settings

**Read in data**
```{r}
#Data source 1- basespace files
        #Set working directory
        wd<- "C:/Users/Jenna/Documents/UKy_RNASeq/RawCounts/"
        
        matureInput<- read.csv(paste(wd, "miRNA_matureHits_downloadedSeparately.csv", sep=""))
        isomirInput<-read.csv(paste(wd, "miRNA_isomirs_downloadedSeparately.csv",sep=""))
        
        #Remove sequence info from isomirs, then collapse isomirs together
        isomirInput$isomirs<- gsub(".*_", "", isomirInput$X)
        isomirInput<-isomirInput[,-1]
        isomir_collapsed<-isomirInput %>% group_by(isomirs) %>%  summarise_each(funs(sum)) %>% as.data.frame()
        rownames(isomir_collapsed)<-isomir_collapsed$isomirs
        isomir_collapsed<- isomir_collapsed[,-1]
        
        #rbind mature and isomirs together. Make sure columns are in the same order
        rownames(matureInput)<-matureInput$X
        matureInput<-matureInput[,-1]
        counts_merged<-rbind(matureInput,isomir_collapsed)
        
        #remove everything after 3rd "-" (-1, -5p, etc), then merge and sum together based on new miR IDs
        counts_collapsedInput<-counts_merged
        counts_collapsedInput$geneIDs<- sub("^([^-]*-[^-]*-[^-]*).*", "\\1", rownames(counts_collapsedInput))
        counts_collapsedInput$geneIDs<- gsub("mir", "miR", counts_collapsedInput$geneIDs)
                
        #Collapse isomirs and mature toegether
        counts_collapsed<-counts_collapsedInput %>% group_by(geneIDs) %>% summarise_each(funs(sum)) %>% as.data.frame
        rownames(counts_collapsed)<-counts_collapsed$geneIDs
        counts_collapsed<- counts_collapsed[,2:ncol(counts_collapsed)]
        rownames(counts_collapsed)<- rownames(counts_collapsed)
        
#Data source 2- command line miRDeep2. Collapsed into 1 file in "ReofrmatingRawmiRDeep2Output.Rmd"
                wd<-"E:/UKy/miRNA_Bams/SmallRNA_092019/"
                CLcountsInput<-read.csv(paste(wd, "AllCountsFromCL_Merged.csv",sep=""), row.names = "geneID" )
                CLcountsInput<- CLcountsInput[,-1]
                CLcounts<-as.data.frame(apply(CLcountsInput, 2, as.integer))
                rownames(CLcounts)<-rownames(CLcountsInput)
                
#Select which data source to use for the rest of the analysis. CL counts used downstream.
counts<-CLcounts
```

#Comparing TTN+ vs TTN- vs Controls
```{r}
#process, rename samples
hits<- counts
colnames(hits)<- gsub("counts_Control", "UK.Control.", gsub("counts_Sample", "X", colnames(hits)))

#get rid of any IDs that have 0 counts for all samples 
hits<- hits[rowSums(hits)>0,]

#Remove sample 4
hits<-hits[,!(colnames(hits)=="X4")]
```

**Read in metadata, reformat**
```{r}
#Read in metadata info
wd<- "C:/Users/Jenna/Documents/UKy_RNASeq/"
meta<- read.csv(paste(wd, "MetaData_AllSamples.csv",sep=""))
rownames(meta)<- meta$Sample
meta<- meta[,-1]
meta<- meta[!(rownames(meta) %in% c(paste("X", c(10,14,4), sep=""))),] 
```


```{r}
detach("package:metaseqR")
detach("package:DESeq")
```

**Plot on PCA- need to create DESeq2 object first**
```{r}
#Create dds object, rundeseq2
dds<-RunDESeq2(hits, meta)

#plot PCA
p<-MyPCA(dds, meta, 20, 14,c(Control="steelblue4", TTNpos="goldenrod2", TTNneg="violetred4"),-5,25,-10,20,5,"sampleGroup" )
```

Samples slightly cluster by CM status, but do NOT cluster based on TTN mutations

**Extract differential expression results on this DESeq2 object- tpos vs tneg vs controls**
```{r}
#Extract results
res_tPosvControl<- ContrastDGE(c("Control", "TTNpos"), dds)
res_tNegvControl<-ContrastDGE(c("Control", "TTNneg"), dds)
res_tPosvtNeg<-ContrastDGE(c("TTNpos", "TTNneg"), dds)


#Normalize
rld<- rlog(dds)

#select top genes: all sig DE miRNAs
topgenes<- unique(c(rownames(res_tPosvControl[res_tPosvControl$padj<0.05,]), 
             rownames(res_tNegvControl[res_tNegvControl$padj<0.05,]),
             rownames(res_tPosvtNeg[res_tPosvtNeg$padj<0.05,])))

#Alt: select top in each, even if not necessarily DE
topn<-15
topgenes<- unique(c(rownames(res_tPosvControl[1:topn,]), 
             rownames(res_tNegvControl[1:topn,]),
             rownames(res_tPosvtNeg[1:topn,])))

#plot heatmap
annotationColors<-list(Group=c(Control="steelblue4", TTNpos="goldenrod2", TTNneg="violetred4"))
PlotHeatmap(rld, meta, topgenes,-.5,1.2,12,annotationColors)
```

#Rerun the above analysis, comparing ALL CM vs controls

**Create DESeq2 object and PCA plot**
```{r}
#Create new metadata
meta_CMvControls<-meta
meta_CMvControls$sampleGroup<- gsub("TTNpos", "CM", 
                                    gsub("TTNneg", "CM", meta_CMvControls$sampleGroup))

#Create DESeq object and run DESeq2
dds_CMvControls<-RunDESeq2(hits, meta_CMvControls)

#Plot PCA
MyPCA(dds_CMvControls, meta_CMvControls, 20, 14,c(Control="steelblue4", CM="olivedrab"),-5,25,-10,20,5,"sampleGroup" )
```

**Extract differential expression results on this DESeq2 object- CM vs Controls**
```{r}
#extract results
res<- as.data.frame(results(dds_CMvControls))
res<- na.omit(res)
res<- res[order(res$padj),]

nrow(res[res$padj<0.05,])#19 DE genes

sig<-res[res$padj<0.05 & abs(res$log2FoldChange>0.5),] #11
miRNA_sig<-res[res$padj<0.05,] #39
summary(results(dds))
```

**Plot Heatmap for these DE results**

```{r}
#Normalize
rld<- rlog(dds_CMvControls)

#select top X DE genes for control vs CM
topgenes<-rownames(sig)
topgenes<-rownames(miRNA_sig)
topgenes<-rownames(res[1:25,])


#plot heatmap
annotationColors<-list(Group=c(Control="steelblue4", CM="olivedrab"))
PlotHeatmap(rld, meta_CMvControls, topgenes,-.5,1.2,12,annotationColors)
```

**Write DE results to csv**
```{r}
write.csv(as.data.frame(res), paste(wd, "DESeq2/miRNADESeq2Results_matureInput.csv", sep =""))
```



#Performing random sample permutation to show (hopefully) that our groups give the highest number of DEGs
```{r}
AllN_DEGs<-list()
All_SampleGroups<-list()
count=1
while (count<=10){
        #Randomly assign each sample to "pos" or "neg"
        meta_random<- meta
        meta_random$sampleGroup<-sample(meta_random$sampleGroup)
     
        #Perform DE using this new metadata
        dds_random<- RunDESeq2(hits,meta_random)
        
        #Extract comparisons for pos vs neg
        DEGs<-ContrastDGE(c("CM", "control"), dds_random)
        SigDEGs<- DEGs[DEGs$padj<0.05,]
        AllN_DEGs[count]<-nrow(SigDEGs)
        All_SampleGroups[[count]]<-meta_random
        
        print(count)
        count=count+1
}
```


```{r}
sessionInfo()
```

